**1.3**

Skriv et program som oppretter en variabel kalt antallStudenter som står for det totale antallet
studenter på et kurs, samt en variabel antallKvinner som står for antallet kvinner. Programmet
skal skrive ut antallet studenter og kvinneandelen på skjermen. Kjør programmet med ulike verdier
for antall studenter totalt og kvinner blant dem.

In [38]:
antallStudenter = 100
antallKvinner = 75

kvinneAndel = antallKvinner / antallStudenter

print(f"Totalt antall studenter: {antallStudenter} , kvinneandel: {kvinneAndel:.2f}")

Totalt antall studenter: 100 , kvinneandel: 0.75


**1.4** 

Lag et program som konverterer fra Fahrenheit til Celsius.

In [39]:
graderFahrenheit = 60
graderCelsius = (graderFahrenheit - 32) * 5/9

print(f"{graderFahrenheit} grader Fahrenheit tilsvarer {graderCelsius} grader Celsius")

60 grader Fahrenheit tilsvarer 15.555555555555555 grader Celsius


Modifiser programmene slik at gradene angis med to desimaler.

In [40]:
graderFahrenheit = 60
graderCelsius = (graderFahrenheit - 32) * 5/9

print(f"{graderFahrenheit} grader Fahrenheit tilsvarer {graderCelsius:.2f} grader Celsius")

60 grader Fahrenheit tilsvarer 15.56 grader Celsius


Spesialtegn for Celsius- og Fahrenheit-grader kan skrives med tegnkoder, henholdsvis
u"\N{degree celsius}" og u"\N{degree fahrenheit}".

In [41]:
graderFahrenheit = 60
graderCelsius = (graderFahrenheit - 32) * 5/9

print(f"{graderFahrenheit}\N{degree fahrenheit} = {graderCelsius:.2f}\N{degree celsius}")

60℉ = 15.56℃


**1.5**

Her ser du tilordninger til 15 variable som holder hhv temperatur, nedbørsmengde og vindstyrke for
mandag til fredag i en uke. Koden er vedlagt i filen værdata.py.

In [42]:
import værdata

ManTemp=10.1
ManNedbør=3.5
ManVind=8.1
TirTemp=12.1
TirNedbør=10.5
TirVind=12.9
OnsTemp=13
OnsNedbør=3.9
OnsVind=12.1
TorTemp=12.9
TorNedbør=0
TorVind=5.4
FreTemp=11.2
FreNedbør=0.1
FreVind=5.2

vekedagar = ["Man", "Tir", "Ons", "Tor", "Fre"]

for dag in vekedagar:
    temp = globals()[dag +'Temp']
    nedbør = globals()[dag +'Nedbør']
    vind = globals()[dag +'Vind']
    print(f"{dag} {temp:>5} {nedbør:>5} {vind:>5}")


Man  10.1   3.5   8.1
Tir  12.1  10.5  12.9
Ons    13   3.9  12.1
Tor  12.9     0   5.4
Fre  11.2   0.1   5.2


Endre programmet ved å legge til en variabel kalt bredde som står for kolonnebredde i
utskriften, dvs hvor mange plasser som brukes på hvert tall. Programmet skal først skrive ut
kolonnebredden og så skrive verdiene som i punkt a) men der bredde-variabelen bestemmer
avstanden mellom kolonnene.

In [43]:
import værdata

ManTemp=10.1
ManNedbør=3.5
ManVind=8.1
TirTemp=12.1
TirNedbør=10.5
TirVind=12.9
OnsTemp=13
OnsNedbør=3.9
OnsVind=12.1
TorTemp=12.9
TorNedbør=0
TorVind=5.4
FreTemp=11.2
FreNedbør=0.1
FreVind=5.2

vekedagar = ["Man", "Tir", "Ons", "Tor", "Fre"]

kolonnebredde = 5
print(f"{kolonnebredde = }")

for dag in vekedagar:
    temp = globals()[dag +'Temp']
    nedbør = globals()[dag +'Nedbør']
    vind = globals()[dag +'Vind']
    print(f"{dag} {temp:>{kolonnebredde}} {nedbør:>{kolonnebredde}} {vind:>{kolonnebredde}}")


kolonnebredde = 5
Man  10.1   3.5   8.1
Tir  12.1  10.5  12.9
Ons    13   3.9  12.1
Tor  12.9     0   5.4
Fre  11.2   0.1   5.2


**1.5** ALTERNATIV

In [44]:
import værdata

bredde='9'
mal='%'+bredde+'.1f'

print('kolonnebredde', bredde)
print('Man',mal%ManTemp,mal%ManNedbør,mal%ManVind)
print('Tir',mal%TirTemp,mal%TirNedbør,mal%TirVind)
print('Ons',mal%OnsTemp,mal%OnsNedbør,mal%OnsVind)
print('Tor',mal%TorTemp,mal%TorNedbør,mal%TorVind)
print('Fre',mal%FreTemp,mal%FreNedbør,mal%FreVind)

kolonnebredde 9
Man      10.1       3.5       8.1
Tir      12.1      10.5      12.9
Ons      13.0       3.9      12.1
Tor      12.9       0.0       5.4
Fre      11.2       0.1       5.2


**2.5**

Lag et program som leser inn radiusen til en sirkel fra tastaturet og beregner arealet til sirkelen.
Skriv ut med tre desimaler.

In [45]:
from math import pi

radius = float(input("Radius: "))
areal = pi*radius**2
print(f"Arealet til en sirkel med radius {radius} er {areal:.3f}")

Arealet til en sirkel med radius 2.0 er 12.566


**2.6**

Les to siffer n og m fra tastaturet, sett de sammen til det tosifrede tallet mn
(dvs tallet som starter på m og slutter på n - ikke produktet m*n) og omvendt nm.
Beregn så kvadratroten av mn*nm. Skriv svaret med to desimaler.

In [46]:
from math import sqrt

n = input("1. siffer: ")
m = input("2. siffer: ")

nm = int(n+m)
mn = int(m+n)
kvadratrot = sqrt(nm*mn)

print(f"Kvardratroten av {nm} * {mn} = {kvadratrot:.2f}")

Kvardratroten av 12 * 21 = 15.87


**2.7**

Sett inn parenteser tilsvarende PEMDAS-reglene i følgende uttrykk.

In [47]:
print(2*3+4**5)
print((2*3)+(4**5))

print(6**5-4/3*2+1)
print(((6**5)-((4/3)*2))+1)

print(1+2+3*4/5*6-7*8)
print(((1+2)+(((3*4)/5)*6))-(7*8))

1030
1030
7774.333333333333
7774.333333333333
-38.6
-38.6


Sett inn parenteser slik at uttrykkene nedenfor blir entydige (dvs at PEMDAS-reglene er
unødvendige) og begge evalueres til samme tall (hint: 21)

In [48]:
(6+5)*4/2-1==((4*(2**2))-3**2)*(3**2-3*2)

True

**2.8**

Lag et program som skriver ut et tilfeldig antall spørsmåltegn på skjermen og ber brukeren gjette
hvor mange det er. Programmet skal skrive ut True dersom gjetningen er riktig, False ellers.
Eksempler (dialogen tar seg best ut på engelsk):

In [49]:
from random import randint

tilfeldigTall = randint(0,30)
print("?"*tilfeldigTall)
gjett = int(input("Guess the number of ?'s: "))

print("That's True!" if gjett == tilfeldigTall else "That's False!")
print(f"{tilfeldigTall} is the right answer.")

????????????????????????????
That's False!
28 is the right answer.


**2.9**

Lag et program som ber brukeren om et tre-sifret tall. Programmet skal skrive ut alle permutasjoner
av sifrene i dette tallet. Dvs alle tall man kan få ved å stokke om på sifrene i tallet.
For et tre-sifret tall vil det være totalt 6 permutasjoner, inkludert det opprinnelige tallet.

In [50]:
tall = [*input("Oppgi et tresifret tall: ")]
premutasjoner = []

premutasjoner.append(tall[0]+tall[1]+tall[2])
premutasjoner.append(tall[0]+tall[2]+tall[1])
premutasjoner.append(tall[1]+tall[0]+tall[2])
premutasjoner.append(tall[1]+tall[2]+tall[0])
premutasjoner.append(tall[2]+tall[0]+tall[1])
premutasjoner.append(tall[2]+tall[1]+tall[0])

print("Permutajoner:", *premutasjoner)

Permutajoner: 123 132 213 231 312 321


**2.10**

Lag et program der du spiller stein-saks-papir mot maskinen.
• Du velger et tall, 1 (papir),2 (saks) eller 3(stein)
• Maskinen velger tilsvarende tall, tilfeldig.
• Bruk sannhetsverdiene True/False til å rapportere om utfallet ble seier, tap eller uavgjort

In [51]:
from random import randint

valg = int(input("Velg (1=papir,2=saks eller 3=stein):"))
maskin = randint(1,3)

seier = False
uavgjort = False

if valg == 1 and maskin == 3 or valg == 2 and maskin == 1 or valg == 3 and maskin == 2:
    seier = True
elif valg == maskin:
    uavgjort = True

print("Maskinen velger:", maskin)
print("Jeg vant:", seier)
print("Uavgjort: ", uavgjort)
print("Maskinen vant: ", not(seier) if uavgjort == False else False)

Maskinen velger: 3
Jeg vant: True
Uavgjort:  False
Maskinen vant:  False


**2.11**

Les inn et hel-tall fra tastaturet og avgjør om det er et oddetall. Svar True eller False

In [52]:
tall = int(input("tall: "))

print(f"{tall} is an odd number:", True if tall%2!=0 else False)

23 is an odd number: True


**2.12**

Les inn to tall n og m (der n£m) fra tastaturet, beregn n+...+m og skriv ut svaret. 

In [53]:
fra = int(input("Fra: "))
til = int(input("Til: "))

sum = 0
i = fra

while i <= til:
    sum += i
    i += 1

print(f"{fra} + ... + {til} = {sum}")

3 + ... + 2 = 0


**2.13**

Eksperimenter deg fram til hvilken presedens % har i PEMDAS-hierarkiet

% har samme presedens som /, *, //

**3.5**

Skuddår forekommer når årstallet er delelig med 4. Unntaket er det første året i århundrer som ikke
er delelig med 400 (for eksempel år 1800 og 1900).
Lag et program som leser inn et årstall og avgjør om det er et skuddår.
(Hint: tallet N er delelig med tallet M dersom N%M==0)

In [54]:
årstall = int(input("årstall: "))

skuddår = False

if årstall % 400 == 0:
    skuddår = True
elif årstall % 4 == 0 and årstall % 100 != 0:
    skuddår = True

print("skuddår" if skuddår == True else "ikke skuddår")

skuddår


**3.6**

Lag et program som ber brukeren taste inn en timelønn og et antall arbeidstimer og skriver ut hvor
mye hen har tjent på arbeidet. Dersom arbeidstimene overstiger 40 skal man få 10%
overtidsbetaling. Dersom brukeren taster inn negative tall, eller verdier som ikke er tall, skal
programmet gi melding om feilen og gi brukeren en sjanse til å oppgi et gyldig tall i stedet

In [55]:
timelønn = None
arbeidstimer = None

while True:
    timelønn = input("Timelønn: ")
    arbeidstimer = input("Arbeidstimer: ")

    try:
        timelønn = float(timelønn)
        arbeidstimer = float(arbeidstimer)

        if timelønn < 0:
            print("Vennligst oppgi positiv timelønn, prøv igjen")
        elif arbeidstimer < 0:
            print("Vennligst oppgi positiv antall arbeidstimer, prøv igjen")
        else:
            break

    except ValueError:
        print("Feil: Vennligst angi tall")

inntekt = arbeidstimer * timelønn

if arbeidstimer > 40:
    inntekt += (arbeidstimer - 40) * 0.1

print(f"Du har tjent {inntekt:.1f} kroner.")


Feil: Vennligst angi tall
Feil: Vennligst angi tall
Feil: Vennligst angi tall
Feil: Vennligst angi tall
Feil: Vennligst angi tall
Feil: Vennligst angi tall
Feil: Vennligst angi tall
Feil: Vennligst angi tall
Feil: Vennligst angi tall
Du har tjent 1.0 kroner.


**3.7**

Lag et program som gitt en bokstav (a-h) og et tall (1-8) avgjør hvilken farge den tilsvarende ruten
har på et sjakkbrett.

In [1]:
bokstaver = "ABCDEFGH"

bokstav = input("bokstav: ").upper()
tall = int(input("tall: "))

if (bokstaver.find(bokstav) % 2 == 0) == (tall % 2 == 0):
    farge = "hvit"
else: farge = "sort"

print("Ruten er", farge)

Ruten er hvit


**3.10**

FrueMafiaen er en kriminell organisasjon der kun kvinner kan være medlemmer. Mafiaen er strengt
hierarkisk med Presidenten som øverste leder. På neste nivå finner vi Kommandanter, deretter
Kapteiner så Sersjanter og nederst finner vi Soldatene. De som ikke er kvalifiserte til opptak, kan
melde seg som Støttemedlem. Man kan stige i gradene, etter søknad, basert på ansiennitet og
meritter. Vedtektene spesifiserer kvalifikasjonskravene til de ulike gradene:

• Minstekrav for alle grader, inkludert støttemedlem, er å være Kvinne, minst 20 år gammel og
ha middels dårlig vandel.

• Soldater må være under 30 år og ikke ha barn

• Det samme gjelder for Sersjanter, som i tillegg må ha minst 4 års ansiennitet og ikke rent
rulleblad.

• Kapteiner må være yngre enn 40 år, ha minst 6 års ansiennitet og ha tiltale eller straffedom
på rullebladet.

• Kommandanter må også være yngre enn 40 år med minst 6 års ansiennitet, men må ha en
straffedom på rullebladet samt ha temmelig dårlig vandel.

• Presidenten må oppfylle samme krav som kommandanter, men i tillegg ha minst 8 års
ansiennitet. Dessuten kan ikke presidenten ha et fornavn som kan forveksles med et 

In [ ]:
class Person:
    def __init__(self, navn, alder, kjønn, antall_barn, vandel, rulleblad, ansiennitet):
        self.navn = navn
        self.alder = alder
        self.kjønn = kjønn
        self.antall_barn = antall_barn
        self.vandel = vandel
        self.rulleblad = rulleblad
        self.ansiennitet = ansiennitet
        self.stillinger = []
    
    def kjønnsnøytraltNavn(self):
        kjønnsnøytrale_navn = ["Kristen", "Kim", "Janne", "Tony"]
        if self.navn in kjønnsnøytrale_navn:
            return True
        return False

    def minstekrav(self):
        if self.kjønn == "K" and self.alder >= 20 and self.vandel < 7:
            self.stillinger.append("Støttemedlem")
            return True
        return False
    
    def soldat(self):
        if self.alder < 30 and self.antall_barn == 0:
            self.stillinger.append("Soldat")
            return True
        return False
    
    def sersjant(self):
        if self.soldat() and self.ansiennitet >= 4 and self.rulleblad != "rent":
            self.stillinger.append("Sersjant")
    
    def kaptein(self):
        if self.alder < 40 and self.ansiennitet > 6 and (self.rulleblad == "tiltalt" or self.rulleblad == "straffet"):
            self.stillinger.append("Kaptein")
            return True
        return False
    
    def kommandant(self):
        if self.kaptein() and self.rulleblad == "straffet" and self.vandel < 4:
            self.stillinger.append("Kommandant")
            return True
        return False
    
    def president(self):
        if self.kommandant() and self.ansiennitet > 8 and not(self.kjønnsnøytraltNavn()):
            self.stillinger.append("President")
            return True
        return False
    
    def sjekk_stillinger(self):
        if self.minstekrav():
            print(f"{self.navn} er kvalifisert for:")

            self.soldat()
            self.sersjant()
            self.kaptein()
            self.kommandant()
            self.president()
            
            for stilling in list(set(self.stillinger)):
                print(stilling)
        else:
            print(f"{self.navn} kvalifiserer ikke for noen stillinger.")

navn = input("Navn: ")
alder = int(input("Alder: "))

kjønn = input("Kjønn (M/K): ").strip().upper()
while kjønn not in ["M", "K"]:
    print("Ugyldig kjønn. Skriv 'M' eller 'K'.")
    kjønn = input("Kjønn (M/K): ").strip().upper()

antall_barn = int(input("Antall barn: "))
vandel = int(input("Vandel, på skala fra 1(dårlig) til 9(plettfri):"))

rulleblad = input("Rulleblad (straffet/tiltalt/mistenkt/rent):").strip().lower()
while rulleblad not in ["straffet", "tiltalt", "mistenkt", "rent"]:
    print("Ugyldig verdi for rulleblad. Velg mellom 'straffet', 'tiltalt', 'mistenkt' eller 'rent'.")
    rulleblad = input("Rulleblad (straffet/tiltalt/mistenkt/rent):").strip().lower()

ansiennitet = int(input("Ansiennitet (år): "))


Person(navn, alder, kjønn, antall_barn, vandel, rulleblad, ansiennitet).sjekk_stillinger()

Kim er kvalifisert for:
Kommandant
Kaptein
Støttemedlem


**4.6**

Skuddår forekommer når årstallet er delelig med 4. Unntaket er det første året i århundrer som ikke
er delelig med 400 (for eksempel år 1800 og 1900).
Lag en funksjon som gitt et årstall avgjør om det er et skuddår. 

In [ ]:
def sjekkSkuddår(år):
    if år % 400 == 0:
        return True
    if år % 4 == 0 and år % 100 != 0:
        return True
    return False
    
sjekkSkuddår(1800)

False

**4.7**

I math-modulen finnes tallet p (math.pi) definert med et visst antall desimaler.

Lag en funksjon pi(d) som returnerer p med d desimaler. Modifiser slik at standardverdi for antallet desimaler være 2, dvs at returverdien blir 3.14 når man kaller pi()

In [ ]:
import math #ville brukt 'from math import pi' om funksjonen ikke skulle hete pi()'

def pi(antall_desimaler=2):
    maks_desimaler = len(str(math.pi))
    mal = '%.' + str(antall_desimaler) + 'f'

    if antall_desimaler > maks_desimaler:
        print("For mange desimaler")
        print(math.pi)
    else:
        print(mal % math.pi)

pi()

3.14


**4.8**

Fullfør funksjonene nyObservasjon som gitt en observert verdi registrerer observasjoner samt
snitt som returnerer snitt-verdien av observasjonene som er registrert

In [ ]:
antallObservasjoner=0
akkumulerteObservasjoner=0

def nyObservasjon(obs):
    global akkumulerteObservasjoner, antallObservasjoner
    akkumulerteObservasjoner += obs
    antallObservasjoner += 1

def snitt():
    if antallObservasjoner == 0:
        return 'Har ingen observasjoner'
    return akkumulerteObservasjoner/antallObservasjoner

print(snitt())
nyObservasjon(10)
nyObservasjon(15)
nyObservasjon(12)
print(snitt())

Har ingen observasjoner
12.333333333333334


**4.9**

Lag en funksjon som trekker et tilfeldig spillkort. Det skal være valgfritt å angi hvilken farge (Spar, Ruter, Kløver eller Hjerter) det skal trekkes fra, ellers skal også fargen trekkes tilfeldig. 

In [ ]:
from random import randint

def tilfeldigKort(type = None):
    verdier = list("23456789JQKA")+["10"]
    typer = ["Spar", "Kløver", "Ruter", "Hjerte"]
    symbol = {"Spar": '\u2660', "Ruter": '\u2666', "Kløver": '\u2663', "Hjerte": '\u2665'}

    tilfeldigType = typer[randint(0,3)] if type == None else type
    tilfeldigVerdi = verdier[randint(0,12)]

    print(symbol[tilfeldigType] + tilfeldigVerdi)

tilfeldigKort("Ruter")


♦5


**4.10**

Lag en funksjon som sorterer tre tall og skriver ut resultate

In [ ]:
def sorter3(tall1,tall2,tall3):
    print(*sorted([tall1, tall2, tall3]))

sorter3(5,9,2)

2 5 9


**4.11**

Lag et program der brukeren kan taste inn ønsket rett (pizza eller calzone) og så skriver ut
instruksjonene for retten. Oppskriftene har fellestrekk, unngå å gjenta disse i programkoden. Lag
egne funksjoner for del-oppgaver i prosessene, som å lage deig og steke osv

In [ ]:
def lag_deig():
    print("bland mel, gjær og vann kna grundig og la heve i 1 time")

def stek(minutter):
    print(f"stek i {minutter} minutter ved 220 grader")

def tilbred_kjøtdeig():
    print("brun kjøttdeig på middels varme")

def tilbred_tomatsaus():
    print("hell over tomatsaus og la det koke opp")

def kjevle_deig():
    print("kjevle ut deigen i en sirkel")

def pizza():
    lag_deig()
    tilbred_kjøtdeig()
    tilbred_tomatsaus()
    kjevle_deig()
    print("fordel fyllet utover hele deigen strø på revet ost")
    stek(15)

def calzone():
    lag_deig()
    tilbred_kjøtdeig()
    tilbred_tomatsaus()
    kjevle_deig()
    print("fordel fyllet på halve deigen brett over den andre halvparten og klem kantene sammen")
    stek(20)


rett = input("Hva vil du ha til middag? ").lower()
if rett == "pizza": pizza()
if rett == "calzone": calzone()



bland mel, gjær og vann kna grundig og la heve i 1 time
brun kjøttdeig på middels varme
hell over tomatsaus og la det koke opp
kjevle ut deigen i en sirkel
fordel fyllet på halve deigen brett over den andre halvparten og klem kantene sammen
stek i 20 minutter ved 220 grader


**4.12**

Her ser du en funksjon som beregner karaktersnitt for et semester som består av 3 emner à 10
studiepoeng.

def semesterSnitt(karakter1,karakter2,karakter3):
 samlet=tilTall(karakter1)+tilTall(karakter2)+tilTall(karakter3)
 snitt=samlet/3
 return tilBokstav(snitt)
 
a) Fullfør funksjonsdefinisjonen ved å lage funksjonene tilTall(bokstav) og tilBokstav(tall)
som hhv regner om fra en bokstav-karakter til et tall og omvendt fra et tall til en bokstav-karakter.
Omregningsregler

• A=5, B=4, C=3, D=2, E=1, F=0

• Tall-karakteren rundes opp til nærmeste bokstav

In [ ]:
from math import ceil

def semesterSnitt(*karakterer):
    samlet = 0
    for karakter in karakterer:
        samlet += tilTall(karakter)
    snitt=samlet/len(karakterer)
    return tilBokstav(snitt)

def tilTall(bokstav):
    bokstav = bokstav.upper()
    return karakter_skala.index(bokstav)

def tilBokstav(tall):
    tall = ceil(tall)
    return karakter_skala[tall]

#lager en liste med karakterer. 
# kunne alternativt: reversed(list("ABCDEF")), men her er det snakk om få elementer
karakter_skala = ["F", "E", "D", "C", "B", "A"]

semesterSnitt('A','C','D','B')

'B'

**4.15**

In [ ]:
def sekundKonvertering(sekunder):
    timer = sekunder // 3600
    minutter = sekunder % 3600 // 60
    sekunder = sekunder % 60

    return f"{timer} time(r), {minutter} minutt(er) og {sekunder} sekund(er)"

sekundKonvertering(500)

'0 time(r), 8 minutt(er) og 20 sekund(er)'

**4.16**

In [2]:
def ord99(tall):
    enkeltSiffer = ["", "en", "to", "tre", "fire", "fem", "seks", "sju", "åtte", "ni", "ti"]
    dobbeltSiffer = ["", "", "tjue", "tretti", "førti", "femti", "seksti", "sytti", "åtti", "nitti"]

    if tall < 10:
        print(enkeltSiffer[tall])
    else:
        tall = str(tall)
        tall_delt = list(tall)
        print(dobbeltSiffer[int(tall_delt[0])]+enkeltSiffer[int(tall_delt[1])])

ord99(21)


tjueen


**4.17**

Lag en funksjon som velger 3 tilfeldige sifre mellom 1 og 9 og
returnerer tallet som består av de tre sifrene i stigende rekkefølge

In [ ]:
from random import randint

def tilfeldig3():
    print(*sorted([randint(1,9) for i in range(3)]),sep="")

tilfeldig3()

178


**5.2**

Lag en funksjon skrivSekvens(sekvens) som skriver ut elementene i en
sekvens på en linje, adskilt av mellomrom.

In [ ]:
def skrivSekvens(input):
    for i in input:
        print(i, end=" ")

skrivSekvens('Python')
skrivSekvens(range(90,100))

P y t h o n 90 91 92 93 94 95 96 97 98 99 

**5.10**

En ekvipasje består av en hest og en rytter. Tenk deg startlister for
ridekonkurranse der de to første elementene utgjør den første ekvipasjen, de to
neste er den 2. ekvipasjen og så videre

In [ ]:
startliste = ['Kari','Blakken','Liv','Dølaprinsen','Eli','Sofus','Gro','Blåne']

def hester(liste):
    return liste[1::2]

def ryttere(liste):
    return liste[::2]

hester(startliste)
ryttere(startliste)

['Blakken', 'Dølaprinsen', 'Sofus', 'Blåne']

**5.11**

Lag en funksjon pyramide(høyde) som gitt et tall tegner en pyramidefigur av
*-tegn med gitt høyde (høyde= antall *'er i midten av figuren)

In [ ]:
def pyramide(høyde):
    #øvre del & midten
    for i in range(1,høyde):
        print('*'*i)

    #nedre del
    for i in range(høyde,1,-1):
        print('*'*i)

pyramide(4)


*
**
***
****
***
**


**5.12**

Lag et program som skriver ut renteutviklingen av et beløp, år for år, inntil det samlede beløpet overstiger en gitt sum

In [ ]:
startbelop = float(input("Startbeløp: "))
rentesats = float(input("Rentesats (%): "))/100
onsketBelop = float(input("Ønsker beløp: "))

belop = startbelop
år = 0

while belop < onsketBelop:
    år += 1
    belop += belop*rentesats

    print(f"År {år} : {belop:.2f}")

År 1 : 1030.00
År 2 : 1060.90
År 3 : 1092.73
År 4 : 1125.51
År 5 : 1159.27
År 6 : 1194.05
År 7 : 1229.87


**5.12**

Lag en funksjon som kan hjelpe dørvakten på et utested å holde styr på hvor
mange gjester som er sluppet inn og sikre at antallet ikke overstiger kapasiteten i
lokalet. Kapasiteten skal være parameter til funksjonen, og vakten skal kunne angi
antall personer i hver gruppe som ankommer eller forlater lokalet.

In [ ]:
def kapasitetVakt(kapasitet):
    print("-- Bruksanvisning --")
    print("Noen ankommer - tast 1")
    print("Noen går - tast 2")
    print("Avslutt - 0")
    print("--------------------")

    ledigKapasitet = kapasitet

    while True:
        print("Ledig kapasitet:", ledigKapasitet)
        valg = int(input("Hva skjer? "))

        if valg == 1:
            antallPersoner = int(input("Hvor mange kommer? "))
            
            if antallPersoner > ledigKapasitet:
                print("For mange! Slipp inn", ledigKapasitet)
                ledigKapasitet = 0
                
            else: ledigKapasitet -= antallPersoner
            
        if valg == 2:
            antallPersoner = int(input("Hvor mange går? "))
            ledigKapasitet += antallPersoner

        if valg == 0:
            print("Takk for nå")
            break
    
kapasitetVakt(10)

-- Bruksanvisning --
Noen ankommer - tast 1
Noen går - tast 2
Avslutt - 0
--------------------
Ledig kapasitet: 10
Ledig kapasitet: 8
Ledig kapasitet: 10
Ledig kapasitet: 5
Ledig kapasitet: 4
Ledig kapasitet: 0
For mange! Slipp inn 0
Ledig kapasitet: 0
Ugyldig antall! Skriv inn et tall.
Ledig kapasitet: 0
Ugyldig valg! Skriv inn et tall.
Ledig kapasitet: 0
Ugyldig valg! Skriv inn et tall.
Ledig kapasitet: 0
Takk for nå


**5.14**

I denne oppgaven skal du lage programmer for å spille Stein-Saks-Papir mot
maskinen over flere runder. To varianter: ‘Best av N runder‘ og ‘Førstemann til N
seiere’ Brukeren skal oppgi et antall runder som skal spilles. Den som vinner flest runder,
vinner sammenlagt.

In [ ]:
from random import randint

runder = int(input("Hvor mange runder? "))

seier = 0
uavgjort = 0
tap = 0

for i in range(runder):
    valg = int(input(f"Runde {i+1} - Velg (1=papir,2=saks eller 3=stein): "))
    maskin = randint(1,3)
    print(f"Maskinen velger {maskin}")

    if valg == maskin:
        print("Det ble uavgjort")
        uavgjort += 1

    elif valg == 1 and maskin == 3 or valg == 2 and maskin == 1 or valg == 3 and maskin == 2:
        print("Det ble seier")
        seier += 1
    
    else:
        print("Det ble tap")
        tap += 1

print("--- Sammenlagt --------------")
print("Du vant" if seier > tap else "Du tapte" if seier < tap else "uavgjort", f"{seier} - {tap} ({uavgjort} uavgjort)")


Maskinen velger 2
Det ble tap
Maskinen velger 3
Det ble tap
--- Sammenlagt --------------
Du tapte 0 - 2 (0 uavgjort)


**5.15**

Lag en funksjon gangetabell som gitt et tall n skriver ut gangetabellen for
tallene fra 1 til n, altså en tabell på formen

In [ ]:
def gangetabell(tall, plasser=3):
    for i in range(1,tall+1):
        for j in range(1,tall+1):
            print(f"{i*j:>{plasser}}", end=" ")
        print()

gangetabell(5,2)

 1  2  3  4  5 
 2  4  6  8 10 
 3  6  9 12 15 
 4  8 12 16 20 
 5 10 15 20 25 


**5.16**

Skriv om histogram-funksjonen fra oppgave 2 slik at histogrammet skrives ut
stående i stedet for liggende.

In [ ]:
def histogram(input):
    for i in range(max(input),0,-1):
        tekst = ""
        for j in range(len(input)):
            if input[j] >= i:
                tekst += "*"
            else: 
                tekst += " "
        print(tekst)

histogram([3,5,7,2,9,8,1,5,0,7])

    *     
    **    
  * **   *
  * **   *
 ** ** * *
 ** ** * *
*** ** * *
****** * *
******** *


**5.17**

Lag en funksjon for et gjette-spill. Gitt et tall N skal det generere et tilfeldig tall
mellom 1 og N og så la brukeren gjette hvilket tall det er. Hvis brukeren gjetter riktig,
skal funksjonen skrive ut "Gratulerer, du gjettet riktig!" og antall forsøk det tok. Hvis
brukeren gjetter feil, skal funksjonen gi hint om tallet er høyere eller lavere, og la
brukeren gjette på nytt. Programmet skal avsluttes når brukeren treffer riktig tall.

In [ ]:
from random import randint

def gjette_spill(maks_verdi):
    mål = randint(1,maks_verdi)
    forsok = 0

    while True:
        print(f"Gjett et tall mellom {1} og {maks_verdi}")
        gjett = int(input("Gjett tallet: "))
        forsok += 1

        if gjett == mål:
            print(f"Gratulerer, du gjettet riktig! Du trengte {forsok} forsøk.")
            break
        
        print("Tallet er", "høyere" if gjett < mål else "lavere")

gjette_spill(100)

Gjett et tall mellom 1 og 100
Tallet er lavere
Gjett et tall mellom 1 og 100
Tallet er høyere
Gjett et tall mellom 1 og 100
Tallet er høyere
Gjett et tall mellom 1 og 100
Tallet er lavere
Gjett et tall mellom 1 og 100
Tallet er lavere
Gjett et tall mellom 1 og 100
Gratulerer, du gjettet riktig! Du trengte 6 forsøk.


**5.18**

Lag et program som skriver ut alle permutasjonene av bokstavene a,b og c. Dvs
alle måter å stokke om ‘abc’ (inkludert ‘abc’ selv). Bruk for-løkker til å generere
permutasjonene.

In [ ]:
from itertools import permutations

ord = 'abc'

for perm in permutations(ord, len(ord)):
    print(''.join(perm))

abc
acb
bac
bca
cab
cba


In [ ]:
def permutasjoner(ord):
    for i in range(len(ord)):
        premutasjon = ord[i]
        
        for j in range(len(ord)):
            if j != i:
                for k in range(len(ord)):
                    if k != i and k != j:
                        premutasjon = ord[i] + ord[j] + ord[k]
                        print(premutasjon)

ord = 'abc'
permutasjoner(ord)


abc
bac
cab


**5.21**

Lag funksjoner som gitt et siffer n (heltall mellom 1 og 9) skriver ut et kvadrat med
sifrene opp til n.

In [ ]:
def tegnKvardrat(tall):
    for i in range(1,tall+1):
        print(f"{i} "*tall)

tegnKvardrat(5)

print()

def tegnKvardrat(tall):
    for i in range(1,tall+1):
        print(*range(1,tall+1))

tegnKvardrat(5)


1 1 1 1 1 
2 2 2 2 2 
3 3 3 3 3 
4 4 4 4 4 
5 5 5 5 5 

1 2 3 4 5
1 2 3 4 5
1 2 3 4 5
1 2 3 4 5
1 2 3 4 5


**5.22**

Lag et program som skriver ut fargene til alle rutene på sjakk-brettet

In [ ]:
rad = [x+1 for x in range(8)]
kolonne = list("ABCDEFGH")

for k in kolonne:
    for r in rad:
        farge = "hvit" if (r % 2 == 0) == (kolonne.index(k) % 2 == 0) else "sort"
        print(k+str(r), ":", farge)

A1 : sort
A2 : hvit
A3 : sort
A4 : hvit
A5 : sort
A6 : hvit
A7 : sort
A8 : hvit
B1 : hvit
B2 : sort
B3 : hvit
B4 : sort
B5 : hvit
B6 : sort
B7 : hvit
B8 : sort
C1 : sort
C2 : hvit
C3 : sort
C4 : hvit
C5 : sort
C6 : hvit
C7 : sort
C8 : hvit
D1 : hvit
D2 : sort
D3 : hvit
D4 : sort
D5 : hvit
D6 : sort
D7 : hvit
D8 : sort
E1 : sort
E2 : hvit
E3 : sort
E4 : hvit
E5 : sort
E6 : hvit
E7 : sort
E8 : hvit
F1 : hvit
F2 : sort
F3 : hvit
F4 : sort
F5 : hvit
F6 : sort
F7 : hvit
F8 : sort
G1 : sort
G2 : hvit
G3 : sort
G4 : hvit
G5 : sort
G6 : hvit
G7 : sort
G8 : hvit
H1 : hvit
H2 : sort
H3 : hvit
H4 : sort
H5 : hvit
H6 : sort
H7 : hvit
H8 : sort


Definer en klasse kalt Person som har attributter for personens navn og alder. 

In [ ]:
class Person:
    def __init__(self, navn, alder):
        self.navn = navn
        self.alder = alder

class Bil:
    def __init__(self, merke, modell, eier):
        self.merke = merke
        self.modell = modell
        self.eier = eier

    def skriv(self):
        return f'{self.merke}, {self.modell}, eier: {self.eier.navn}'

kari=Person('Kari Normann',35)
per=Person('Per Spellmann',67)
bil1=Bil('Volvo','EX90',kari)
bil2=Bil('Toyota','Yaris',per)

bil1.skriv()


'Volvo, EX90, eier: Kari Normann'

In [ ]:
class Person:
    def __init__(self, navn, alder):
        self.navn = navn
        self.alder = alder

class Bil:
    def __init__(self, merke, modell, navn, alder):
        self.merke = merke
        self.modell = modell
        self.eier = Person(navn, alder)

    def skriv(self):
        return f'{self.merke}, {self.modell}, eier: {self.eier.navn}'

bil3=Bil('Kia','E-Soul','Ola Normann',29)
bil3.skriv()
bil3.eier.alder

29

**6.7**

Dørvakten til et selskapslokale skal holde styr på hvor mange gjester som er sluppet inn og sikre at
antallet ikke overstiger kapasiteten i lokalet. Lag en klasse kapasitetVakt for å holde telling med
antall personer i et lokale, med metoder for å angi antall personer som kommer og går, samt for å
hente ledig kapasitet i lokalet. Vakten skal potensielt vokte flere lokaler, av varierende størrelse,
samtidig. 

In [ ]:
class kapasitetVakt:
    def __init__(self, kapasitet):
        self.kapasitet = kapasitet

    def ledig(self):
        return self.kapasitet

    def kommer(self, antall):
        if antall > self.kapasitet:
            print("Slipp inn", self.kapasitet)
            self.kapasitet = 0
        else: self.kapasitet -= antall

    def går(self, antall):
        self.kapasitet += antall

lokale1=kapasitetVakt(10)
lokale2=kapasitetVakt(15)
lokale1.kommer(4)
lokale2.kommer(5)
lokale2.kommer(2)
print(lokale1.ledig())
print(lokale2.ledig())
lokale1.kommer(9)
print(lokale1.ledig())
lokale1.går(5)
print(lokale1.ledig())

6
8
Slipp inn 6
0
5


**6.10**

Fibonacci-serien er en tallrekke der de to første tallene i serien er 1’ere og deretter er det neste
tallet summen av de to forrige: 1, 1, 2, 3, 5, 8, 13, 21, … osv.
En fibonacci-generator er et objekt x som holder et tall i fibonacci-serien og med en metode
x.neste() som oppdaterer til det neste tallet i serien og returnerer det

In [ ]:
class Fibonacci:
    def __init__(self):
        self.neste_tall = 1
        self.tall = 1
        self.siste = 1
    
    def neste(self):
        self.neste_tall = self.siste + self.tall 
        self.siste = self.tall
        self.tall = self.neste_tall
        return self.neste_tall

# Eksempel på bruk
g = Fibonacci()
print(g.neste()) 
print(g.neste()) 
print(g.neste()) 
print(g.neste()) 


2
3
5
8


**6.12**

Lag et program som kan skrive ut en kjede av forgjengere til en gitt person. (foreldre, besteforeldre,
oldeforeldre, osv i vilkårlig mange generasjoner). Kjeden skal spesifiseres ved en sekvens som består
av F’er og M’er, dvs en tekst som dette: ‘FMMF’. I dette tilfellet skal programmet skrive ut personens
far (F), så hans mor (M) etterfulgt av hennes mor (M) og til slutt hennes far (F). Hint: lag en person-
klasse som registrerer navn samt far og mor - som også er personer!. På den måten kan du lage en
struktur av nestede objekter som representerer stamtreet bakover fra en person.

In [ ]:
class Person:
    def __init__(self, navn, far = None, mor=None):
        self.navn = navn
        self.far = far
        self.mor = mor

def forgjengere(person, sekvens):
    for i in sekvens:
        if i == "M":
            if person.mor:
                person = person.mor
                print(person.navn)
            else:
                print("Kjeden er for lang")
        if i == "F":
            if person.far:
                person = person.far
                print(person.navn)
            else:
                print("Kjeden er for lang")

anders = Person("Anders")
anna = Person("Anna")
tor = Person("Tor")
guri = Person("Guri")
kjell = Person("Kjell")
olaug = Person("Olaug")
arne = Person("Arne")
solveig = Person("Solveig")
jens = Person("Jens", anders, anna)
liv = Person("Liv", tor, guri)
petter = Person("Petter", kjell, olaug)
sara = Person("Sara", arne, solveig)
ola = Person("Ola", jens, liv)
kari = Person("Kari", petter, sara)
per = Person("Per", ola, kari)
        
forgjengere(per,'MMF')


Kari
Sara
Arne


**7.5**

Lag en funksjon som gitt et fullt navn vil skrive det ut med etternavnet
først, deretter fornavnet, evt etterfulgt av forbokstaven i alle mellomnavn

In [ ]:
def skrivNavn(navn):
    navn = navn.split()
    fornavn = navn[0]
    etternavn = navn[-1]
    mellomnavn = navn[1:-1]
    mellomnavn_formatert = [x[0] + '.' for x in mellomnavn]

    return f"{etternavn}, {fornavn} {" ".join(mellomnavn_formatert)}"

skrivNavn('Ole Martin Eek Olsen')

'Olsen, Ole M. E.'

**7.6**

Lag en funksjon som tar inn en liste av tekststrenger samt et heltall N som argumenter og
returnerer en ny liste med de av tekststrengene som har en lengde på minst N tegn.

In [ ]:
def lengre_enn(liste,lengde):
    return [ord for ord in liste if len(ord)>=lengde]

liste = ['Vått', 'og', 'kaldt', 'og', 'breiflabb', 'overalt']
lengre_enn(liste,7)

['breiflabb', 'overalt']

**7.7**

Lag en funksjon som skriver ut teksten til sangen 'Tre små musikanter'
der man kan velge å erstatte vokalene i teksten med en oppgitt vokal som opsjon

In [ ]:
sang = """tre små musikanter på høybro plass\n
sto og spilte på en kontrabass\n
så kom en konstabel, spurte hva var hendt\n
tre små musikanter på høybro plass"""

def treSmåMusikanter(vokal = None):

    vokaler = list("aeiouyøæå")
    revidert_sang = ""
    sang = """tre små musikanter på høybro plass sto 
    og spilte på en kontrabass 
    så kom en konstabel, spurte hva var hendt 
    tre små musikanter på høybro plass"""

    if not(vokal): return sang

    for bokstav in sang:
        if bokstav in vokaler:
            revidert_sang += vokal
        else:
            revidert_sang += bokstav
    
    return revidert_sang

treSmåMusikanter("i")

'tri smi misikintir pi hiibri pliss sti \n    ig spilti pi in kintribiss \n    si kim in kinstibil, spirti hvi vir hindt \n    tri smi misikintir pi hiibri pliss'

**7.8**

Lag en funksjon som, gitt en liste av kongelige på formen 'Tittel Navn',
skriver ut navnet (uten tittel) til alle prinser og prinsesser.
Ta høyde for at titlene kan skrives med stor eller liten stor forbokstav.

In [ ]:
kongefamilie = ['Kong Otto', 'dronning Magda', 'Kronprins Ulf',
'kronprinsesse Frida', 'prinsesse Angela', 'Prins Peter',
'prinsesse Petra', 'grevinne Gunda', 'Grev Wilhelm',
'Prins Ingolf']

def prinserOgPrinsesser(liste):
    prinserOgPrinsesser_liste = [person.split(maxsplit=1)[1] for person in liste if person.lower().startswith("prinsesse") or person.lower().startswith("prins")]
   
    for person in prinserOgPrinsesser_liste:
        print(person)

prinserOgPrinsesser(kongefamilie)

Angela
Peter
Petra
Ingolf


**7.9**

Lag en funksjon som gitt en tekst returnerer teksten reversert

In [ ]:
def snu(ord):
    return ord[::-1]

snu('Python')

'nohtyP'

**7.10**

Lag en funksjon som sjekker om et ord er et palindrom. Dvs at det leses likt forlengs og
baklengs

In [ ]:
def palindrom(ord):
    return ord == ord[::-1]

palindrom('regning')

False

**7.11**

Lag en funksjon som tar inn en tekststreng og sjekker hvor mange ganger hvert tegn
forekommer i strengen. For hvert tegn skal funksjonen skrive ut antallet ganger det
forekommer i teksten.

In [ ]:
def tell_tegn(ord):
    tegn_fortegnelse = {}

    for tegn in ord:
        if tegn not in tegn_fortegnelse:
            tegn_fortegnelse[tegn] = 0
        tegn_fortegnelse[tegn] += 1
    
    for tegn, antall in tegn_fortegnelse.items():
        print(f"{tegn} : {antall}")

tell_tegn('programmering')

p : 1
r : 3
o : 1
g : 2
a : 1
m : 2
e : 1
i : 1
n : 1


**7.12**

Lag en funksjon mest_brukte_ord som tar en setning/streng og returnerer det mest
brukte ordet i den setningen. Hvis det er flere ord som forekommer flest ganger er det
vilkårlig hvilket som blir returnert.

In [ ]:
def mest_brukte_ord(streng):
    mest_brukt = ""
    antall_ganger_brukt = 0

    for ord in streng.split():
        if streng.count(ord) > antall_ganger_brukt:
            mest_brukt = ord
            antall_ganger_brukt = streng.count(ord)
    
    return mest_brukt

mest_brukte_ord('Dette er en test som er enkel og grei')

'er'

**7.13**

Lag en funksjon som gitt en tekststreng og et tall N returnerer listen med alle delstrenger
(utsnitt) av lengde N.

In [ ]:
def delstrenger(streng, lengde):
    liste = []

    for i in range(len(streng)-(lengde-1)):
        liste.append(streng[i:i+lengde])
    
    return liste

delstrenger('Vått og kaldt',4)

['Vått',
 'ått ',
 'tt o',
 't og',
 ' og ',
 'og k',
 'g ka',
 ' kal',
 'kald',
 'aldt']

**7.15**

Lag en funksjon flett(t1,t2) som fletter tekstene t1 og t2 sammen til 1 tekst
Dvs den skal returnere en tekst som starter med det 1.tegnet i t1 etterfulgt av 1. tegnet i t2.
Så kommer 2.tegn i t1, så 2. tegn i t2 og så videre. Dersom t1 og t2 ikke er like lange skal
resten fra den lengste legges til slutt i resultatet

In [ ]:
def flett(t1, t2):
    min_lengde = min(len(t1), len(t2))
    i = 0
    string = ""

    while i < min_lengde:
        string += t1[i]+t2[i]
        i += 1

    if len(t1) > len(t2): string += t1[i:]
    if len(t1) < len(t2): string += t2[i:]

    return string

flett('abcdefgh', '12345')

'a1b2c3d4e5fgh'

**7.16**

Cæsar skift er en enkel form for kryptering av ord der hver bokstav erstattes av bokstaven
som ligger N plasser lengre ut i alfabetet. Hvis det ikke er nok bokstaver lengre utover
fortsetter man fra begynnelsen av alfabetet.
Nå skal du lage funksjoner for kryptering med N=1, altså at hver bokstav erstattes med den
neste i alfabetet, og at 'å' erstattes med 'a'.
Du kan ta for gitt at input-ordet ikke inneholder tall, mellomrom eller andre tegn enn de
norske bokstavene fra a-å.

In [ ]:
def krypter(tekst, n):
    kryptert_ord = ""
    for bokstav in tekst:
        kryptert_ord += chr(ord(bokstav)+n).lower()

    return kryptert_ord

krypter('abcøæå', 1)

'bcdùçæ'

In [ ]:
def krypter(tekst, n=1):
    alfabet = list("abcdefghijklmnopqrstuvwxyzæøå")
    kryptert_ord = ""

    for bokstav in tekst:
        plassering = alfabet.index(bokstav)+n if alfabet.index(bokstav)+n < len(alfabet) else alfabet.index(bokstav)+n-len(alfabet)
        kryptert_ord += alfabet[plassering]

    return kryptert_ord

krypter('programmering')

'qsphsbnnfsjoh'

In [ ]:
def dekrypter(tekst, n):
    alfabet = list("abcdefghijklmnopqrstuvwxyzæøå")
    dekryptert_ord = ""

    for bokstav in tekst:
        plassering = alfabet.index(bokstav)-n if alfabet.index(bokstav)-n > 0 else alfabet.index(alfabet[alfabet.index(bokstav)-n])
        dekryptert_ord += alfabet[plassering]

    return dekryptert_ord

dekrypter('qsphsbnnfsjoh', 1)

'programmering'

**7.17**

Bytt tekstverdier i to variable uten å innføre en ekstra helpevariabel.

In [ ]:
a='hei'
b='hopp'
print(a,b)
a= a+b
b= a[0:a.index(b)]
a= a[len(b):]
print(a,b)

hei hopp
hopp hei


In [ ]:
a='hei'
b='hopp'
print(a,b)
a, b = b, a
print(a,b)

hei hopp
hopp hei


In [ ]:
def subliste(sub, liste):
    for i in liste:
        if liste[i:i+len(sub)] == sub:
            return True
    return False

subliste([3,4,5], [1,2,3,4,5,6,7])
subliste([3,4,5], [7,6,5,4,3,2,1])


False

**8.3**

Last ned ordlistefilen 'ordliste.txt' fra kursets hjemmeside på MittUib og lagre den der du lagrer
python-programmene dine. Lag et program som leser inn ordlisten og lar brukeren taste inn norske
søkeord og svarer med den engelske oversettelsen. Prøv å finne alle oppslag i ordlisten som starter
på brukerens søkeord.

In [ ]:
print("Oppgi søkeord (avslutt med 'slutt')")

while True:
    søkeord = input("Søkeord: ").strip()
    if søkeord.lower() == "slutt": break

    try:
        with open('ordliste.txt', 'r') as fil:
            funnet = False
            for ord in fil:
                ord = ord.strip()
                
                if ',' in ord:
                    try:
                        norsk, engelsk = ord.split(',', maxsplit=1)
                        if norsk.startswith(søkeord):
                            print(f"{norsk} = {engelsk}")
                            funnet = True
                    except ValueError:
                        pass 
            if not funnet:
                print("Ingen treff funnet.")
    except FileNotFoundError:
        print("Filen 'ordliste.txt' ble ikke funnet. Sørg for at den eksisterer.")


Oppgi søkeord (avslutt med 'slutt')
tilkobling (filer) =  file handle
tekstsegment =  slice
tekstverdi =  string
tilordning =  assignment
tolk =  interpreter
tom funksjon =  void function
tom streng =  empty string
traversering =  traversal
tuppel =  tuple
tilkobling (filer) =  file handle
tilordning =  assignment


**8.4**

Lag en funksjon erstattOrd(filnavn, gammeltOrd, nyttOrd) som erstatter alle forekomster
av ordet gammeltOrd med nyttOrd i filen.

In [ ]:
def erstattOrd(filnavn, gammeltOrd, nyttOrd):
    tekst = ""
    with open(filnavn, 'r') as fil:
        for linje in fil:
            if gammeltOrd in linje:
                linje = linje.replace(gammeltOrd, nyttOrd)
            tekst += linje

    with open(filnavn, 'w') as fil:
        fil.write(tekst)

erstattOrd('treSmåMusikanter.txt','høybro', 'danmarks')

**8.5**

- telefon.txt der personer er registrert med navn og telefonnummer, i tilfeldig rekkefølge.

- adresse.txt der personer er registrert med navn og adresse. Denne filen er sortert alfabetisk

Adresse- og telefonlistene inneholder ikke nødvendigvis de samme personene.
Lag et program som tar utgangspunkt i personene i telefonlisten og lager en ny fil med
kontaktinformasjon for disse. Alle skal være registrert med telefon samt adresse der den er kjent

In [ ]:
telefon_liste = {}

with open('telefon.txt', 'r') as fil:
    for linje in fil:
        linje = linje.strip()
        navn, nummer = linje.split()
        telefon_liste[navn] = nummer

adresse_liste = {}
with open('adresse.txt', 'r') as fil:
    for linje in fil:
        linje = linje.strip()
        navn, adresse = linje.split(maxsplit=1)
        adresse_liste[navn] = adresse

with open('kontaktinfo.txt', 'w') as fil:
    for navn, nummer in telefon_liste.items():
        if navn in adresse_liste:
            fil.write(f"{navn} {nummer} {adresse_liste[navn]}\n")
        else:
            fil.write(f"{navn} {nummer}\n")

**8.6**

Vi ser videre på den sorterte adresselisten adresse.txt
Lag en funksjon settinn(person, fil)som kan sette inn en ny person/adresse, på riktig sted i
filen så den forblir sortert.

In [ ]:
def settinn(data, filnavn):
    ny_fil = ""
    plassert = False

    with open(filnavn, 'r') as fil:
        for linje in fil:
            if linje > data and plassert == False: 
                ny_fil += data + "\n"
                plassert = True
            ny_fil += linje

        if plassert == False:
            ny_fil += data

    with open(filnavn, 'w') as fil:
        fil.write(ny_fil)

settinn('Siv Kirkegata 1', 'adresse.txt')

**8.10**

Lag en funksjon som erstatter alle sifre i en fil med tilsvarende tallord, resultatet legges i en ny fil.

In [ ]:
def erstattSifre(filnavn):
    tall_i_bokstaver = ["null", "en", "to", "tre", "fire", "fem", "seks", "sju", "åtte", "ni"]
    ny_fil = ""

    with open(filnavn, 'r') as fil:
        for person in fil:
            person = person.strip()
            navn, nummer = person.split()
            nummer_til_bokstaver = [tall_i_bokstaver[int(siffer)] for siffer in nummer]
            ny_fil += f'{navn}   {"-".join(nummer_til_bokstaver)}\n'

    with open('NYTTtelefon.txt', 'w') as fil:
        fil.write(ny_fil)

erstattSifre('telefon.txt')

**9.1.5**

Liste-metoden L.remove(e) fjerner kun 1 forekomst av e fra listen L.
Lag en funksjon fjernAlle(Liste, e) som fjerner alle forekomster av e slik at

a) funksjonen endrer listen permanent.

b) funksjonen returnerer en resultat-liste mens originalen forblir uendret.

In [ ]:
def fjernAlle(liste, element):
    antall = liste.count(element)
    for i in range(antall):
        liste.remove(element)
    return liste

L=[1,2,3,4,5,6,5,4,3,2,1]
fjernAlle(L,4)

[1, 2, 3, 5, 6, 5, 3, 2, 1]

In [ ]:
def fjernAlle(liste, element):
    antall = liste.count(element)
    liste = liste[:]
    for i in range(antall):
        liste.remove(element)
    return liste

L=[1,2,3,4,5,6,5,4,3,2,1]
fjernAlle(L,4)
print(L)
print(fjernAlle(L,4))

[1, 2, 3, 4, 5, 6, 5, 4, 3, 2, 1]
[1, 2, 3, 5, 6, 5, 3, 2, 1]


**9.1.6**

Lag en funksjon som tar inn to lister som argumenter. Den skal returnere en ny liste som
inneholder elementene som er felles for begge listene, uten duplikater.

In [ ]:
def felles_elementer(l1, l2):
    l3 = []
    for element in l1:
        if element in l2 and element not in l3:
            l3.append(element)
    return l3

felles_elementer([1, 2, 3, 4, 3, 5], [4, 5, 5, 6, 7, 8])

[4, 5]

**9.1.8**

L1 er en sub-liste av L2 dersom L1 er et segment i L2. Dvs at elementene fra L1 forekommer i L2,
uavbrutt og i samme rekkefølge.

a) Lag en funksjon som sjekker om en liste er subliste av en annen liste

b) Lag en funksjon som finner alle posisjonene til en subliste i en større liste

In [ ]:
def subliste(sub, liste):
    for i in range(len(liste)):
        if i+len(sub) <= len(liste) and liste[i:i+len(sub)] == sub:
            return True
        elif i+len(sub) > len(liste) and liste[i:] == sub:
            return True
    return False

subliste([3,4,6], [1,2,3,4,5,6,7])

False

In [ ]:
def sublistePos(sub, liste):
    pos = []
    for i in range(len(liste)):
        if i+len(sub) <= len(liste) and liste[i:i+len(sub)] == sub:
            pos.append(i)
        elif i+len(sub) > len(liste) and liste[i:] == sub:
            pos.append(i)
    return pos

sublistePos([4,5], [1,2,3,4,5,6,5,4,3,4,5,4,5,6,3,4,5,6,7])

[3, 9, 11, 15]

**9.1.9**

 Lag funksjoner for å håndtere 2-dimensjonale tabeller med tall, representert som lister

Funksjonen lagTabell(antallKolonner, antallRader) skal opprette en tabell med de gitte
dimensjonene. Initielt kan alle celler inneholde verdien 0.

In [ ]:
def lagTabell(antallKolonner, antallRader):
    tabell = []
    for i in range(antallRader):
        tabell.append([])
        for j in range(antallKolonner):
            tabell[i].append("0")
    return tabell

def skrivTabell(tabell):
    rader = len(tabell)
    for rad in range(rader):
        print(*tabell[rad]) #evt " ".join(rad)
    print()

def fyllTabell(tabell, liste):
    i = 0
    rader = len(tabell)
    for rad in range(rader):
        kolonner = len(tabell[rad])
        for kolonne in range(kolonner):
            tabell[rad][kolonne] = liste[i]
            i += 1
    
def hentVerdi(tabell, rad, kolonne):
    return tabell[rad-1][kolonne-1]

def settVerdi(tabell, rad, kolonne, verdi):
    tabell[rad-1][kolonne-1] = verdi

def leggTilRad(tabell, liste, pos):
    tabell.insert(pos-1, liste)

t=lagTabell(3,4)
fyllTabell(t, [11,12,13,21,22,23,31,32,33,41,42,43])
skrivTabell(t)
leggTilRad(t, [1,1,1], 3)
skrivTabell(t)

11 12 13
21 22 23
31 32 33
41 42 43

11 12 13
21 22 23
1 1 1
31 32 33
41 42 43



**9.1.10**

Lag en funksjon som tar inn en liste av strenger og returnerer en ny liste der strengene er sortert
alfabetisk, uten å bruke en innebygget Python-funksjon (som sort() eller sorted()).

In [ ]:
def sorter_strenger(liste):
    sortertliste = liste[:]

    for i in range(len(liste)):
        antall_større = 0
        for j in range(len(liste)):
            if liste[i] > liste[j]:
                antall_større += 1
        sortertliste[antall_større] = liste[i]
    
    return sortertliste


strenger = ["Python", "Java", "C++", "Ruby"]
sorter_strenger(strenger)

['C++', 'Java', 'Python', 'Ruby']

**9.2.6**

Lag en funksjon FibonacciFortegnelse(n) som returnerer fortegnelsen med de n første
Fibonacci-tallene. (Husk: Fibonacci-serien starter slik: 1, 1, 2, 3, 5, 8…. og generelt er hvert tall
summen av de to foregående).

In [ ]:
def FibonacciFortegnelse(lengde):
    fortegnelse = {0: 1}
    for i in range(lengde):
        if i > 1:
            fortegnelse[i+1] = fortegnelse[i] + fortegnelse[i-1]
        elif i > 0:
            fortegnelse[i+1] = fortegnelse[i] + 1
        else:
            fortegnelse[i+1] = 1

    return fortegnelse

FibonacciFortegnelse(5)
for x in FibonacciFortegnelse(8).values():print(x)

1
1
2
3
5
8
13
21
34


**9.2.7**

Du har to fortegnelser A og B der verdiene i A er av samme type som nøklene i B.

In [ ]:
dager={1:'man',2:'tir', 3:'ons', 4:'tor', 5:'fre',6:'lør', 7:'søn'}
temperaturer={'man':11, 'tir':9, 'ons':7, 'tor':12, 'fre':11, 'lør':9, 'søn':8}

def koble(f1, f2):
    f3 = {}
    for x, y in f1.items():
        if y in f2:
            f3[x] = f2[y]
    return f3

koble(temperaturer, dager)

{'ons': 'søn'}

**9.2.8**

I et sosialt nettverk kan en person følge andre personer. Dette er ikke nødvendigvis gjensidig.
En følge-fortegnelse har personers navn som nøkler knyttet til lister av navn på personene de følger.

In [ ]:
følgere = {'Anne':['Berit','Karl','Dag'], 'Berit':['Karl','Ina'], 'Karl':['Anne'], 'Dag':['Anne', 'Ina','Berit'], 'Ina':['Karl','Berit','Dag']}

def gjensidige_følgere(følgefortegnelse, navn):
    gjensidig_liste = []

    if navn in følgefortegnelse: følger = følgefortegnelse[navn]
    else: return f'ukjent person: {navn}'

    for person in følger:
        if person in følgere and navn in følgefortegnelse[person]:
            gjensidig_liste.append(person)

    return gjensidig_liste if gjensidig_liste else "ingen gjensidige følgere"

gjensidige_følgere(følgere, 'Anne')
gjensidige_følgere(følgere, 'Berit')
gjensidige_følgere(følgere, 'Jens')

'ukjent person: Jens'

**9.2.9**

Filen vær.txt inneholder vær-data for en uke.

Lag en funksjon som leser en slik fil og returnerer en nestet fortegnelse

In [ ]:
def lesVær(filnavn):
    f_vær = {}

    with open(filnavn, 'r') as fil:
        overskrift = fil.readline()

        for linje in fil:
            linje = linje.strip()
            dag, temp, nedbør, vind = linje.split()
            f_vær[dag] = {'Temp':temp,'Nedbør':nedbør,'Vind':vind}

    return f_vær

vær=lesVær('vær.txt')

vær['Tir']['Temp']
vær['Fre']['Vind']

for dag, vær_data in vær.items():
    if float(vær_data["Nedbør"]) < 0.3:
        print(dag)

Tor
Fre
Lør


**9.2.10**

Tenk deg nå en nestet fortegnelse der ulike sosiale nettverk er knyttet til sine respektive følge-
fortegnelser

Lag en funksjon følgere(nettverksfortegnelse, sosialt_nettverk,person)som returnerer listen
med de som personen følger i det gitte sosiale nettverket. Ta høyde for at ikke alle personer deltar i
alle sosiale nettverk. Ta høyde for at ikke alle personer deltar i alle sosiale nettverk.

In [ ]:
nettverk=\
{'Facebook':{'Anne':['Berit','Karl','Dag'],'Berit':['Karl','Ina'],'Karl':['Anne'],
'Dag':['Anne', 'Ina','Berit'], 'Ina':['Karl','Berit','Dag']},
'Instagram':{'Anne':['Ina','Eva','Dag'], 'Berit':['Dag'], 'Gro':['Ina'],
'Eva':['Anne','Dag'], 'Dag':['Anne', 'Ina','Berit'],
'Ina':['Anne','Berit','Dag']},
'Threads':{'Jon':['Berit','Karl','Dag'], 'Berit':['Karl','Ina'],
'Åse':['Dag','Jon'],'Karl':['Anne'], 'Dag':['Jon', 'Ina','Berit'],
'Ina':['Karl','Berit','Dag']} }

def følgere(nettverksfortegnelse, sosialt_nettverk, person):
    if person not in nettverksfortegnelse[sosialt_nettverk]: return f'ukjent person: {person}'
    
    return nettverksfortegnelse[sosialt_nettverk][person] if nettverksfortegnelse[sosialt_nettverk][person] else f"{person} følger ingen"

følgere(nettverk,'Facebook','Dag')

['Anne', 'Ina', 'Berit']

Lag en funksjon alle_fulgte(nettverksfortegnelse, person) som skriver ut navn på alle
sosiale nettverk som personen er på, samt navn på alle hen følger på minst ett av nettverkene.
Unngå duplikater.

In [ ]:
def alle_fulgte(nettverk, navn):
    person_nettverk = []
    person_følger = set()

    for platform, brukere in nettverk.items():
        if navn in brukere: 
            person_nettverk.append(platform)
            for følger in brukere[navn]:
                person_følger.add(følger)

    if person_nettverk:
        return f"{navn} er på disse nettverkene: {", ".join(person_nettverk)} og følger disse personene: {", ".join(person_følger)}"

alle_fulgte(nettverk, 'Anne')
alle_fulgte(nettverk, 'Jon')

'Jon er på disse nettverkene: Threads\n og følger disse personene: Berit, Dag, Karl'

**9.2.12**

Lag en funksjon som skriver ut en fortegnelse sortert med hensyn på nøkkelverdien.

In [ ]:
def sorter(fortegnelse):
    return sorted(fortegnelse.items())

engelskeSiffer={3:'three', 4:'four', 5:'five',
0:'zero',1:'one', 2:'two',
6:'six', 7:'seven',8:'eight',9:'nine'}

sorter(engelskeSiffer)

[(0, 'zero'),
 (1, 'one'),
 (2, 'two'),
 (3, 'three'),
 (4, 'four'),
 (5, 'five'),
 (6, 'six'),
 (7, 'seven'),
 (8, 'eight'),
 (9, 'nine')]

In [ ]:
def sorter(fortegnelse):
    return sorted(fortegnelse.items(), key=lambda x: (x[1][1], x[0]))

engelskeSiffer={3:["three", "abc"], 4:["three", "bcd"], 5:["three", "abc"],
0:["three", "ghb"],1:["three", "abc"], 2:['acd', 'three'],
6:["three", "øæå"], 7:["three", "hjk"],8:["three", "oiu"],9:["three", "ab"]}

sorter(engelskeSiffer)

[(9, ['three', 'ab']),
 (1, ['three', 'abc']),
 (3, ['three', 'abc']),
 (5, ['three', 'abc']),
 (4, ['three', 'bcd']),
 (0, ['three', 'ghb']),
 (7, ['three', 'hjk']),
 (8, ['three', 'oiu']),
 (2, ['acd', 'three']),
 (6, ['three', 'øæå'])]

Lag en funksjon som gitt en en-til-en-fortegnelse, oppretter den inverse fortegnelsen der nøklene
har blitt verdier, og omvendt.

In [ ]:
engelskeSiffer={3:'three', 4:'four', 5:'five',
0:'zero',1:'one', 2:'two',
6:'six', 7:'seven',8:'eight',9:'nine'}

def invers(fortegnelse):
    invers_fortegnelse = {}

    for nøkkel, verdi in fortegnelse.items():
        invers_fortegnelse[verdi] = nøkkel

    return invers_fortegnelse

invers(engelskeSiffer)

{'three': 3,
 'four': 4,
 'five': 5,
 'zero': 0,
 'one': 1,
 'two': 2,
 'six': 6,
 'seven': 7,
 'eight': 8,
 'nine': 9}

Anta nå at du kun har den inverse fortegnelsen, og ikke originalen. Skriv ut den inverse
fortegnelsen sortert med hensyn på verdien, dvs skriv ut zero 0 først osv.

In [ ]:
def skrivSortert(fortegnelse):
    for nøkkel, verdi in sorted(fortegnelse.items(), key=lambda x:x[1]):
        print(nøkkel, verdi)

engelskeSiffer = {'three': 3,'four': 4,'five': 5,'zero': 0,'one': 1,'two': 2,'six': 6,'seven': 7,'eight': 8,'nine': 9}
skrivSortert(engelskeSiffer)

zero 0
one 1
two 2
three 3
four 4
five 5
six 6
seven 7
eight 8
nine 9


**10.1.3**

En bedrift har registrert sine ansatte og hvilke kompetanser de har i følgende mengder

In [ ]:
personer={'Liv','Jan','Ida','Gro','Per',
'Dag','Ole','Ine','Tor','Åse'}
programmering={'Per','Åse','Gro','Ida','Dag','Tor'}
matematikk={'Gro','Dag'}
databaser={'Liv','Per','Dag','Ine'}
KI={'Åse','Dag','Ida','Gro','Jan'}
logikk={'Jan', 'Gro', 'Ine'}
UX={'Ine','Tor','Liv'}
systemutvikling={'Åse','Liv','Per','Jan'}
filosofi={'Ole'}


#hvor mange som har kompetanse i systemutvikling
print(len(systemutvikling))

#hvem som har kompetanse i UX eller systemutvikling
print(UX | systemutvikling)

#hvem som har kompetanse i KI og logikk
print(KI & logikk)

#hvem som har kompetanse i KI, men ikke i logikk
print(KI - logikk)

#hvem som har kompetanse i verken KI eller logikk
print(personer - KI - logikk)

#hvem som har kompetanse i databaser eller systemutvikling, men ikke i begge
print((databaser | systemutvikling) - (databaser & systemutvikling))

#om alle som kan logikk også kan programmering (True eller False)
print(logikk in programmering)

#om noen kan både logikk og UX (True eller False)
print(len(UX & logikk) > 0)

4
{'Tor', 'Per', 'Åse', 'Ine', 'Jan', 'Liv'}
{'Gro', 'Jan'}
{'Ida', 'Dag', 'Åse'}
{'Tor', 'Per', 'Ole', 'Liv'}
{'Dag', 'Jan', 'Åse', 'Ine'}
False
True


Lag en funksjon som gitt et personnavn returnerer mengden av kompetanser som personen har og
bruk funksjonen til å skrive ut kompetansene til hver person

In [ ]:
personer={'Liv','Jan','Ida','Gro','Per',
'Dag','Ole','Ine','Tor','Åse'}
programmering={'Per','Åse','Gro','Ida','Dag','Tor'}
matematikk={'Gro','Dag'}
databaser={'Liv','Per','Dag','Ine'}
KI={'Åse','Dag','Ida','Gro','Jan'}
logikk={'Jan', 'Gro', 'Ine'}
UX={'Ine','Tor','Liv'}
systemutvikling={'Åse','Liv','Per','Jan'}
filosofi={'Ole'}

kompetanser = {'programmering':programmering,
'matematikk':matematikk,
'databaser':databaser,
'KI':KI,
'logikk':logikk,
'UX':UX,
'systemutvikling':systemutvikling,
'filosofi':filosofi}

for person in personer:
    person_kompetanser = []
    for tittel, kompetanse in kompetanser.items():
        if person in kompetanse:
            person_kompetanser.append(tittel)
    print(f"{person}: {", ".join(person_kompetanser)}")

Tor: programmering, UX
Per: programmering, databaser, systemutvikling
Ida: programmering, KI
Ine: databaser, logikk, UX
Åse: programmering, KI, systemutvikling
Gro: programmering, matematikk, KI, logikk
Ole: filosofi
Dag: programmering, matematikk, databaser, KI
Jan: KI, logikk, systemutvikling
Liv: databaser, UX, systemutvikling


Lag en funksjon som gitt et team, i form av en mengde av personer, returnerer mengden av
kompetanser som teamet mangler.

In [ ]:
personer={'Liv','Jan','Ida','Gro','Per',
'Dag','Ole','Ine','Tor','Åse'}
programmering={'Per','Åse','Gro','Ida','Dag','Tor'}
matematikk={'Gro','Dag'}
databaser={'Liv','Per','Dag','Ine'}
KI={'Åse','Dag','Ida','Gro','Jan'}
logikk={'Jan', 'Gro', 'Ine'}
UX={'Ine','Tor','Liv'}
systemutvikling={'Åse','Liv','Per','Jan'}
filosofi={'Ole'}

kompetanser = {'programmering':programmering,
'matematikk':matematikk,
'databaser':databaser,
'KI':KI,
'logikk':logikk,
'UX':UX,
'systemutvikling':systemutvikling,
'filosofi':filosofi}

def manglerKompetanse(team):
    manglende_kompetanse = set()
    
    for tittel, kompetanse in kompetanser.items():
        if len(team & kompetanse) == 0:
            manglende_kompetanse.add(tittel)
    
    return manglende_kompetanse

manglerKompetanse({'Liv','Jan'})

{'filosofi', 'matematikk', 'programmering'}

**10.1.5**

Lag en funksjon som kan koble to mengder med tupler. Det vil si at du har to mengder A og B der
andre-komponentene A er av samme type som første-komponentene i B.

In [ ]:
dager={(1,'man'),(2,'tir'),(3,'ons'),(4,'tor'),
(5,'fre'),(6,'lør'),(7,'søn')}

temperaturer={('man',11),('tir',9),('ons',7),('tor',12),
('fre',11),('lør',9),('søn',8)}

def koble(m1, m2):
    m3 = set()
    for (x1, z1) in m1:
        for (z2, x2) in m2:
            if z1 == z2:
                m3.add((x1, x2))
    return m3

koble(dager,temperaturer)
koble(temperaturer,dager)

{('ons', 'søn')}

In [ ]:
l = [1,2,3,4]

siste = l.pop()
første = l.pop(0)

l.insert(0, siste)
l.insert(len(l), første)

print(l)

[4, 2, 3, 1]


In [ ]:
liste = [1, 2, 3, 4, 5]

if len(liste) >= 2:
    liste[0], liste[-1] = liste[-1], liste[0]
print(liste)

**10.2.4**

Lag en funksjon som gitt en liste returnerer listen av alle par som kan dannes fra elementene i
listen.

In [ ]:
def alleKombinasjoner(liste):
    kombinasjoner = []
    for i in liste:
        for j in liste:
            kombinasjoner.append((i, j))

    return kombinasjoner

alleKombinasjoner([1,2,3,4])

[(1, 1),
 (1, 2),
 (1, 3),
 (1, 4),
 (2, 1),
 (2, 2),
 (2, 3),
 (2, 4),
 (3, 1),
 (3, 2),
 (3, 3),
 (3, 4),
 (4, 1),
 (4, 2),
 (4, 3),
 (4, 4)]

**10.2.5**

Gjensyn med kobling av data (jfr. oppgave 7 i lab-oppgavene om fortegnelser)

Du har to lister av par, kalt A og B. Disse kan kobles til en ny liste av par C.
Spesifikt: hvis (x,y) er i A og (y,z) er i B så skal (x,z) være i C.

In [ ]:
dager=[(1,'man'),(2,'tir'),(3,'ons'),(4,'tor'),(5,'fre'),
(6,'lør'),(7,'søn')]
temperaturer=[('man',11),('tir',9),('ons',7),('tor',12),
('fre',11),('lør',9),('søn',8)]

def koble(l1, l2):
    l3 = []
    for x,y1 in l1:
        for y2,z in l2:
            if y1 == y2:
                l3.append((x,z))
    return l3


koble(dager, temperaturer)
koble(temperaturer, dager)

[('ons', 'søn')]

**10.2.6**

Ta utgangspunkt i filer med temperaturdata.

Lag en funksjon som kan lese inn temperaturdata fra en fil og lagre dem som tripler (3-tupler)
i en global liste tempData.

In [ ]:
def lesFil(filnavn):
    global tempData
    tempData = []

    with open(filnavn, 'r') as fil:
        for linje in fil:
            linje = linje.strip()
            sted, dag, temp = linje.split()
            tempData.append((sted, dag, temp))

lesFil('tempFil.txt')
tempData

[('Bergen', 'man', '7.0'),
 ('Bergen', 'ons', '10.0'),
 ('Bergen', 'tir', '9.0'),
 ('Bodø', 'man', '2.0'),
 ('Bodø', 'ons', '5.0'),
 ('Bodø', 'tir', '5.0'),
 ('Kristiansand', 'man', '16.0'),
 ('Kristiansand', 'ons', '13.0'),
 ('Kristiansand', 'tir', '15.0'),
 ('Oslo', 'man', '8.0'),
 ('Oslo', 'ons', '8.0'),
 ('Oslo', 'tir', '5.0'),
 ('Stavanger', 'man', '9.0'),
 ('Stavanger', 'ons', '7.0'),
 ('Stavanger', 'tir', '11.0'),
 ('Tromsø', 'man', '3.0'),
 ('Tromsø', 'ons', '2.0'),
 ('Tromsø', 'tir', '5.0'),
 ('Trondheim', 'man', '8.0'),
 ('Trondheim', 'ons', '8.0'),
 ('Trondheim', 'tir', '9.0')]

Lag en funksjon som lar en bruker taste inn temperaturdata

In [ ]:
def lesFil(filnavn):
    global tempData
    tempData = []

    with open(filnavn, 'r') as fil:
        for linje in fil:
            linje = linje.strip()
            sted, dag, temp = linje.split()
            tempData.append((sted, dag, temp))

def settTemperatur():
    global tempData

    while True:
        try:
            sted = input("Sted: ")
            dag = input("Dag: ")
            temp = float(input("Temperatur: "))
            tempData.append((sted, dag, temp))
            break

        except ValueError:
            print("Ugyldig input, prøv på nytt")

lesFil('tempFil.txt')
tempData
settTemperatur()
tempData


Ugyldig input, prøv på nytt


[('Bergen', 'man', '7.0'),
 ('Bergen', 'ons', '10.0'),
 ('Bergen', 'tir', '9.0'),
 ('Bodø', 'man', '2.0'),
 ('Bodø', 'ons', '5.0'),
 ('Bodø', 'tir', '5.0'),
 ('Kristiansand', 'man', '16.0'),
 ('Kristiansand', 'ons', '13.0'),
 ('Kristiansand', 'tir', '15.0'),
 ('Oslo', 'man', '8.0'),
 ('Oslo', 'ons', '8.0'),
 ('Oslo', 'tir', '5.0'),
 ('Stavanger', 'man', '9.0'),
 ('Stavanger', 'ons', '7.0'),
 ('Stavanger', 'tir', '11.0'),
 ('Tromsø', 'man', '3.0'),
 ('Tromsø', 'ons', '2.0'),
 ('Tromsø', 'tir', '5.0'),
 ('Trondheim', 'man', '8.0'),
 ('Trondheim', 'ons', '8.0'),
 ('Trondheim', 'tir', '9.0'),
 ('1', '1', 1.0)]

lag en funksjon som skriver ut snitt-temperaturen for hvert sted

In [ ]:
def lesFil(filnavn):
    global tempData
    tempData = []

    with open(filnavn, 'r') as fil:
        for linje in fil:
            linje = linje.strip()
            sted, dag, temp = linje.split()
            tempData.append((sted, dag, temp))


def snittSteder():
    steder = set([x for (x,y,z) in tempData])
    for sted_1 in steder:
        temp_liste = []
        for (sted, dag, temp) in tempData:
            if sted_1 == sted: temp_liste.append(float(temp))
        snitt_temperatur = sum(temp_liste)/len(temp_liste)
        print(f"{sted_1} {snitt_temperatur:.2f}")

lesFil('tempFil.txt')
tempData
snittSteder()

Bergen 8.67
Stavanger 9.00
Kristiansand 14.67
Bodø 4.00
Tromsø 3.33
Trondheim 8.33
Oslo 7.00


In [ ]:
def lesFil(filnavn):
    global tempData
    tempData = {}

    with open(filnavn, 'r') as fil:
        for linje in fil:
            linje = linje.strip()
            sted, dag, temp = linje.split()
            tempData[(sted, dag)] = temp

lesFil('tempFil.txt')
tempData

{('Bergen', 'man'): '7.0',
 ('Bergen', 'ons'): '10.0',
 ('Bergen', 'tir'): '9.0',
 ('Bodø', 'man'): '2.0',
 ('Bodø', 'ons'): '5.0',
 ('Bodø', 'tir'): '5.0',
 ('Kristiansand', 'man'): '16.0',
 ('Kristiansand', 'ons'): '13.0',
 ('Kristiansand', 'tir'): '15.0',
 ('Oslo', 'man'): '8.0',
 ('Oslo', 'ons'): '8.0',
 ('Oslo', 'tir'): '5.0',
 ('Stavanger', 'man'): '9.0',
 ('Stavanger', 'ons'): '7.0',
 ('Stavanger', 'tir'): '11.0',
 ('Tromsø', 'man'): '3.0',
 ('Tromsø', 'ons'): '2.0',
 ('Tromsø', 'tir'): '5.0',
 ('Trondheim', 'man'): '8.0',
 ('Trondheim', 'ons'): '8.0',
 ('Trondheim', 'tir'): '9.0'}

**10.2.7**

En ekvipasje består av en hest og en rytter. Tenk deg startlister for ridekonkurranse der de to
første elementene utgjør den første ekvipasjen, de to neste er den 2. ekvipasjen og så videre, dvs
lister på formen
[ rytter1, hest1, rytter2, hest2, ..., rytterN,hestN ]
Lag funksjonen deltakere(startliste) som returnerer et par med listene som består av hhv
hesten og rytterne i startlisten,.

In [60]:
startliste = ['Kari', 'Blakken', 'Liv', 'Dølaprinsen', 'Eli', 'Sofus', 'Gro',
'Blåne']

def deltakere(startliste):
    ryttere = startliste[::2] 
    hester = startliste[1::2]
    return (ryttere, hester)

resultat = deltakere(startliste)
print("Ryttere:", resultat[0])
print("Hester:", resultat[1])


Ryttere: ['Kari', 'Liv', 'Eli', 'Gro']
Hester: ['Blakken', 'Dølaprinsen', 'Sofus', 'Blåne']


In [62]:
startliste = ['Kari', 'Blakken', 'Liv', 'Dølaprinsen', 'Eli', 'Sofus', 'Gro', 'Blåne']

def deltakere(liste):
    ryttere = liste[::2]
    hester = liste[1::2]

    return (hester, ryttere)

(hester,ryttere)=deltakere(startliste)
hester

['Blakken', 'Dølaprinsen', 'Sofus', 'Blåne']

**10.2.9**

Skriv en funksjon sorter_produkter(produkter) som tar inn en liste med produkter i form av
par, der hvert parinneholder produktnavn og pris.

Funksjonen skal sortere produktene basert på pris, fra lavest til høyest, og returnere den sorterte
listen. Hvis to varer har samme pris, skal de sorteres etter navn.

In [63]:
produkter=[('Banan', 10), ('Eple', 8), ('Appelsin', 15), ('Druer',8), ('Pære',10)]

def sorter_produkter(liste):
    return sorted(liste, key=lambda x: (x[1], x[0]))

sorter_produkter(produkter)

[('Druer', 8), ('Eple', 8), ('Banan', 10), ('Pære', 10), ('Appelsin', 15)]

**10.2.11**

Lag en funksjon som tar inn to tupler og returnerer et nytt tuppel som fletter de to tuplene. Det vil si
at resultat-tuppelet skal ha verdier som er hentet vekselsvis fra de to argument-tuplene. Dersom det
ene tuppelet er lengre enn det andre, legges de resterende elementene til slutten.

In [68]:
def flett(t1, t2):
    min_len = min(len(t1), len(t2))
    liste = []

    for i in range(min_len):
        liste.append(t1[i])
        liste.append(t2[i])

    flettet = tuple(liste)
    return flettet + t1[min_len:] + t2[min_len:]

flett((1, 2, 3), ('a', 'b', 'c', 'd', 'e'))

(1, 'a', 2, 'b', 3, 'c', 'd', 'e')

**11.1**

Lag en applikasjon (et brukerorientert program) med et grafisk brukergrensesnitt (GUI).
Brukergrensesnittet skal ha en knapp merket 'Farvel' .

In [1]:
from tkinter import *


vindu = Tk()
knapp = Button(vindu, text="Farvel", command=vindu.destroy)
knapp.grid(row=0, column=0)

vindu.mainloop()

2024-11-28 14:37:25.481 Python[35681:10185241] +[IMKClient subclass]: chose IMKClient_Modern
2024-11-28 14:37:25.481 Python[35681:10185241] +[IMKInputSession subclass]: chose IMKInputSession_Modern


: 

**10.2**

Lag en applikasjon med et grafisk brukergrensesnitt (GUI) som har én enkelt knapp. Til å begynne
med er knappen merket 0, men hver gang knappen trykkes skal verdien økes med 1.

In [4]:
from tkinter import *

vindu = Tk()

n = IntVar()
n.set(0)

def øk():
    n.set(n.get()+1)
    knapp.config(text=str(n.get()))

knapp = Button(vindu, text=str(n.get()), command=øk, width=15)
knapp.grid(column=0, row=0)

vindu.mainloop()

**10.3**

Lag et GUI-program som har et tekstfelt og en knapp. Når brukeren skriver inn tekst i tekstfeltet og
trykker på knappen, skal programmet vise teksten i et etikettfelt (Label).

In [1]:
from tkinter import *

vindu = Tk()

def vis_tekst():
    resultat.config(text=inntastingsfelt.get())

inntastingsfelt = Entry(vindu, width=15)
vis_tekst_knapp = Button(vindu, text="Vis tekst", command=vis_tekst)
resultat = Label(vindu, text="")

inntastingsfelt.grid(column=0, row=0)
vis_tekst_knapp.grid(column=0, row=1)
resultat.grid(column=0, row=2)

vindu.mainloop()

2024-11-28 14:57:04.569 Python[35936:10202781] +[IMKClient subclass]: chose IMKClient_Modern
2024-11-28 14:57:04.569 Python[35936:10202781] +[IMKInputSession subclass]: chose IMKInputSession_Modern


**10.4**

Lag et GUI-program som teller antall avgitte stemmer en avstemning. Programmet skal ha en knapp for hvert alternativ, og en 'Resultat'
-etikett som viser hvor mange stemmer hvert alternativ
har. Programmet skal lagre antall stemmer for hvert alternativ i en liste.

In [3]:
from tkinter import *

stemmer = [0,0,0]

def stemme(alt):
    stemmer[alt] += 1
    resultat.config(text=f"Resultat: 1:{stemmer[0]}, 2:{stemmer[1]}, 3:{stemmer[2]}")
    
vindu = Tk()

for i in range(3):
    Button(vindu, text=f"Alternativ {i+1}", command=lambda i=i:stemme(i)).grid(column=0, row=i)

resultat = Label(vindu, text="Resultat: 1:0, 2:0, 3:0")
resultat.grid(column=0, row=3)

vindu.mainloop()

: 

In [10]:
str1 = 'Ault\'Kelly'
print(str1)

print(type(0xFF))

Ault'Kelly
<class 'int'>


In [17]:
type(0x10)

int

In [18]:
x = 75
def myfunc():
    x = x + 1
    print(x)

myfunc()
print(x)

UnboundLocalError: cannot access local variable 'x' where it is not associated with a value

In [43]:
with open('test.txt', 'r') as fil:
    for linje in fil:
        linje = linje.strip()
        *a, b = linje.split()
        print(' '.join(a),b)

bok 1 1903
bok 2000
bok 3 2102
bok 3 1234


In [48]:
"welcome to the beautiful world of python".capitalize()

'Welcome to the beautiful world of python'

In [51]:
sampleList = [10, 20, 30, 40]
del sampleList[0:6]
print(sampleList)

[]


In [56]:
l1 = [1, 2, 2]
l2 = [1, 2]
print(max(l1,l2))

[1, 2, 2]


In [58]:
[None] * 3

[None, None, None]

In [64]:
keys = ['Ten', 'Twenty', 'Thirty']
values = [10, 20, 30]

fortegnelse = {}
for (x,y) in zip(keys, values):
    fortegnelse[x] = y

print(fortegnelse)

{'Ten': 10, 'Twenty': 20, 'Thirty': 30}


In [1]:
dict1 = {'Ten': 10, 'Twenty': 20, 'Thirty': 30}
dict2 = {'Thirty': 30, 'Fourty': 40, 'Fifty': 50}

dict1 = {**dict1, **dict2}

print(dict1)

{'Ten': 10, 'Twenty': 20, 'Thirty': 30, 'Fourty': 40, 'Fifty': 50}


In [2]:
dict1 = {'Ten': 10, 'Twenty': 20, 'Thirty': 30}
dict2 = {'Thirty': 30, 'Fourty': 40, 'Fifty': 50}

dict1.update(dict2)

print(dict1)

{'Ten': 10, 'Twenty': 20, 'Thirty': 30, 'Fourty': 40, 'Fifty': 50}


In [7]:
for i in range(5):
    print((str(i+1) + " ") * (i+1))

1 
2 2 
3 3 3 
4 4 4 4 
5 5 5 5 5 


In [15]:
'%5d'%123
#5 er plasser totalt

'  123'

In [16]:
'%.2f'%123 
#2 er desimaler

'123.00'

In [17]:
'%7.2f'%123 
#7 er antall plasser totalt
#2 er desimaler

' 123.00'

In [18]:
num = 2

for x in [num*x for x in range(1,11)]:
    print(x)

2
4
6
8
10
12
14
16
18
20


In [25]:
numbers = [12, 75, 150, 180, 145, 525, 50]

for num in numbers:
    if num > 500:
        break
    if num > 150:
        continue
    if num % 5 == 0:
        print(num)

75
150
145


In [26]:
list1 = [10, 20, 30, 40, 50]

for x in list1[::-1]:
    print(x)

50
40
30
20
10


In [30]:
list1 = [10, 20, 30, 40, 50]

for x in reversed(list1):
    print(x)

50
40
30
20
10


In [32]:
for i in range(-10,0):
    print(i)

-10
-9
-8
-7
-6
-5
-4
-3
-2
-1


In [33]:
for i in range(5):
    print(i)
else:
    print("Done!")

0
1
2
3
4
Done!


In [ ]:
num1, num2 = 0, 1

#loop 10 ganger
for i in range(10):
    #print neste tal i serien
    print(num1, end=" ")
    #legg saman siste to verdier til neste tall
    res = num1 + num2
    #oppdater verdier
    num1 = num2
    num2 = res

In [48]:
a = (3,)

a[0]

3

In [52]:
liste = [1,2,3,4,5]

liste.remove(6) if 6 in liste else None
print(liste)

[1, 2, 3, 4, 5]


In [60]:
dict1 = {'Ten': 10, 'Twenty': 20, 'Thirty': 30}
dict2 = {'Thirty': 30, 'Fourty': 40, 'Fifty': 50}

dict1 = {**dict1, **dict2}

print(dict1)

{'Ten': 10, 'Twenty': 20, 'Thirty': 30, 'Fourty': 40, 'Fifty': 50}


In [66]:
dict1 = {'Ten': 10, 'Twenty': 20, 'Thirty': 30}
dict2 = {'Thirty': 30, 'Fourty': 40, 'Fifty': 50}

dict1.update(dict2)
dict3 = dict1

print(dict3)

{'Ten': 10, 'Twenty': 20, 'Thirty': 20, 'Fourty': 40, 'Fifty': 50}


In [68]:
s = set((1,2,3,4,5,6,7,8,9))
for x in s:
    print(x)

1
2
3
4
5
6
7
8
9


**11.5**

Lag en GUI for å skrive verset Tre Små Musikanter der alle vokaler er skiftet ut med en valgt vokal.

In [1]:
from tkinter import *

valg = list("aeiouyæøå")

def sang(vokal=None):
    sang_tekst = ""
    with open('treSmåMusikanter.txt', 'r') as fil:
        for linje in fil:
            for v in valg:
                if vokal:
                    linje = linje.replace(v, vokal)
            sang_tekst += linje
    return sang_tekst

def bytt():
    if vokal.get() == "n":
        sangtekst.set(sang())
    else:
        sangtekst.set(sang(vokal.get()))
vindu = Tk()

valg_panel = Frame(vindu)
valg_panel.grid(column=0, row=0)

vokal=StringVar()
vokal.set("n")

sangtekst = StringVar()
sangtekst.set(sang())

normal = Radiobutton(valg_panel, text= "normal", value="n", variable=vokal)
normal.grid(column=0, row=0)

for i in range(len(valg)):
    Radiobutton(valg_panel, text=valg[i], value=valg[i], variable= vokal, command=bytt).grid(column=i+1, row=0)

etikett = Label(vindu, textvariable=sangtekst)
etikett.grid(column=0, row = 1)

avslutt = Button(vindu, text="Avslutt", command=vindu.destroy)
avslutt.grid(column=0, row=2)

vindu.mainloop()

2024-11-29 17:43:21.756 Python[46434:10892863] +[IMKClient subclass]: chose IMKClient_Modern
2024-11-29 17:43:21.756 Python[46434:10892863] +[IMKInputSession subclass]: chose IMKInputSession_Modern


In [1]:
from tkinter import *

vokaler = list("aeiouyøæå")

def sang(vokal=None):
    sangtekst = ""
    with open("treSmåMusikanter.txt", "r") as fil:
        for linje in fil:
            if vokal:
                for v in vokaler:
                    linje = linje.replace(v, vokal)
            sangtekst += linje
    return sangtekst

def bytt():
    if vokal.get() == "n":
        tekst.set(sang())
    else:
        tekst.set(sang(vokal.get()))

vindu = Tk()

tekst = StringVar()
tekst.set(sang())

vokal = StringVar()
vokal.set("n")

ramme = Frame(vindu)
ramme.grid(column=0, row=0)

normal = Radiobutton(ramme, text="Normal", value="n", variable=vokal, command=bytt)
normal.grid(column=0, row=0)

for i in range(len(vokaler)):
    Radiobutton(ramme, text=vokaler[i], value=vokaler[i], variable=vokal, command=bytt).grid(column=i+1, row=0)

etikett = Label(vindu, textvariable=tekst)
etikett.grid(column=0, row=1)

avslutt = Button(vindu, text="Avslutt", command=vindu.destroy)
avslutt.grid(column=0, row=2)

vindu.mainloop()

2024-11-29 18:05:16.392 Python[46708:10912362] +[IMKClient subclass]: chose IMKClient_Modern
2024-11-29 18:05:16.392 Python[46708:10912362] +[IMKInputSession subclass]: chose IMKInputSession_Modern


: 

In [24]:
x = 10

if x > 9 and x > 8 and x < 5:
    print("Ja")
else:
    print("Nei")

Nei


In [25]:
type(None)

NoneType

In [29]:
x = 1

def f():
    global x
    x += 1
    return x

f()

2

In [30]:
class A:
    en=1
    to=None
    def __init__(self,to): self.to=to
    def skriv(self): print(self.en, self.to)

class B(A):
    tre=3
    def __init__(self):self.to=222
    def skriv(self):print(self.tre, self.to, self.en)
    
a1=A(2)
a2=A(22)
b1=B()
b1.tre=3
L=[a1,a2,b1]
for o in L: o.skriv()
a2.en=11
A.en=4
L=[a1,a2,b1]
for o in L: o.skriv()

1 2
1 22
3 222 1
4 2
11 22
3 222 4


In [32]:
x = 1

def test():
    global x
    x += 1
    return x

test()
x +=1

def test2():
    print(x)
    x += 1

test2()

print(x)

3
4


In [33]:
fortegnelse_temaer = {}

for id, tema_liste in temaer.items():
    for tema in tema_liste:
        if tema not in fortegnelse_temaer:
            fortegnelse_temaer[tema] = {}
        if bøker[id]['utgitt'] not in fortegnelse_temaer[tema]:
            fortegnelse_temaer[tema][bøker[id]['utgitt']] = 0 #kan ikkje være liste
        fortegnelse_temaer[tema][bøker[id]['utgitt']] += 1

sorterte_temaer = dict(sorted(fortegnelse_temaer.items()))
for tema, årstall_dict in sorterte_temaer.items():
    for årstall in årstall_dict:
        print(f"{tema}, {årstall}: {årstall_dict[årstall]}")

NameError: name 'temaer' is not defined

In [42]:
fortegnelse = {("ba123", "år1"):1, ("bbsefdefdawda", "år2"):2}

for x,y in fortegnelse:
    #print(x,y, fortegnelse[(x,y)])
    pass

liste = []

for x in fortegnelse:
    liste.append(x)

liste = sorted(liste)

for x,y in liste:
    print(x,y, fortegnelse[(x,y)])


ba123 år1 1
bbsefdefdawda år2 2


In [3]:
"t e st ".capitalize()
"t e st ".title()

'T E St '

**11.9**

En DNA-sekvens er en streng med forekomster av bokstavene A, T, C og G. Lag en GUI der brukeren
skriver inn en DNA-sekvens og når hen trykker på en knapp merket 'Tell' skal antallet forekomster av
hver bokstav vises i vinduet.

In [ ]:
from tkinter import *

def tell():
    innhold = tekst.get().strip()
    unike_tegn = list(set(innhold))
    data = ""

    for tegn in sorted(unike_tegn):
        antall = innhold.count(tegn)
        data += f"{tegn}:{antall} "
    
    label.config(text=data)

Main = Tk()

tekst = StringVar()

entry = Entry(Main, textvariable=tekst, width = 15)
entry.grid(column=0, row=0)

button = Button(Main, text="Tell", command=tell)
button.grid(column=0, row=1)

label = Label(Main)
label.grid(column=0, row=2)

Main.mainloop()

2024-12-01 09:49:53.286 Python[2224:58359] +[IMKClient subclass]: chose IMKClient_Modern
2024-12-01 09:49:53.286 Python[2224:58359] +[IMKInputSession subclass]: chose IMKInputSession_Modern


: 

**12.1**

Filen representanter.json (vedlagt) inneholder opplysninger om stortingsrepresentanter
(2021), komiteer på stortinget og hvilke representanter som er medlemmer i komiteen

In [8]:
import json

def representanter(fylke):
    with open("representanter.json", 'r') as fil:
        fil = json.load(fil)
        for representant_data in fil["representanter"].values():
            if representant_data["fylke"] == fylke:
                print(f"{representant_data["fornavn"]} {representant_data["etternavn"]} ({representant_data["parti"]})")

representanter("Troms")

Cecilie Myrseth (Arbeiderpartiet)
Kent Gudmundsen (Høyre)
Per-Willy Amundsen (Fremskrittspartiet)
Sandra Borch (Senterpartiet)
Martin Henriksen (Arbeiderpartiet)
Torgeir Knag Fylkesnes (Sosialistisk Venstreparti)


In [1]:
[2,3,4] in [1,2,3,4,5] 

False

In [2]:
[0][0] == 0 

True

In [4]:
len(range(1,10)) == 10 

False

In [5]:
[1,2,3]*2 == [2,4,6] 

False

In [6]:
juba='juba'*2
print(juba+juba)


jubajubajubajuba


In [7]:
juba='juba juba'.split()
print('juba'.join(juba))


jubajubajuba


In [8]:
juba={'juba':'jubajuba', 'jubajuba':'breiflabb', 'breiflabb':'overalt'}
print(juba['jubajuba'])


breiflabb


In [9]:
karakterer={'Erik':'C','Ida':'B','Anne':'B','Ola':'A'}
for (x,y) in sorted(karakterer.items()): print(y, end=' ')


B C B A 

In [11]:
slår={'stein':'saks','saks':'papir','papir':'stein'}
Ola=['saks','stein','saks','papir','stein','stein']
Liv=['stein','saks','papir','papir','stein','saks']
print(slår[Ola[3]]==Liv[3])


False


In [13]:
slår={'stein':'saks','saks':'papir','papir':'stein'}
Ola=['saks','stein','saks','papir','stein','stein']
Liv=['stein','saks','papir','papir','stein','saks']
i=0
for o in Ola:
    if slår[o]== Liv[i]: print(i, end='')
    i=i+1


125

In [15]:
def f(x):
    y=x+x
    return y
x=[1,'1',11,'11']
for y in x: print(f(y), end='+')


2+11+22+1111+

In [16]:
for x in 'abc':
    for y in 'dcb':
        if x>y:print(x+y,end='+')


cb+

In [17]:
A={1,2,3,4,5}
B={4,5,6,7}
print(A|B, A&B, A-B)


{1, 2, 3, 4, 5, 6, 7} {4, 5} {1, 2, 3}


In [7]:
A={1,2,3,4,5}
B={4,5,6,7}
print( (B&(A-B)) | (B-(A&B)) )


{6, 7}


In [5]:
def siste(liste):
    return liste[-1] #-1 vil være det siste elementet i listen

siste(['kaffe','sukker','ost','brød'])

'brød'

In [6]:
def felles(ord1, ord2):
#bruker mengdebygger til å lage en mengde av
#bokstavene som fremgår i begge og unngå duplikater.
    felles_bokstaver = {bokstav for bokstav in ord1 if bokstav in ord2}
    return list(felles_bokstaver) #omgjør til liste


felles('programmering', 'matematikk')

['a', 'i', 'e', 'm']

In [22]:
class Parti:
    def __init__(self, parti_kode, parti_navn):
        self.parti_kode = parti_kode.upper() #sikrer at parti_koden er i store bokstaver
        self.parti_navn = parti_navn.capitalize() #sikrer at første bokstav erstor og resten små
        self.valgår = []
    
    def leggTilÅr(self, årstall):
        if årstall not in self.valgår: #sjekker om valgår allerede er registert
            self.valgår.append(årstall)

    def skriv(self):
        print(f'{self.parti_navn} ({self.parti_kode}) valgår:{self.valgår}')
        
venstre=Parti('V','Venstre')
venstre.skriv()
venstre.leggTilÅr(2013)
venstre.leggTilÅr(2017)
venstre.skriv()

Venstre (V) valgår:[]
Venstre (V) valgår:[2013, 2017]


In [3]:
import json

class Parti:
    def __init__(self, parti_kode, parti_navn):
        self.parti_kode = parti_kode.upper() #sikrer at parti_koden er i store bokstaver
        self.parti_navn = parti_navn.capitalize() #sikrer at første bokstav erstor og resten små
        self.valgår = []
    
    def leggTilÅr(self, årstall):
        if årstall not in self.valgår: #sjekker om valgår allerede er registert
            self.valgår.append(årstall)

    def skriv(self):
        print(f'{self.parti_navn} ({self.parti_kode}) valgår:{self.valgår}')
    
    def json(self):
        fortegnelse = {'partikode': self.parti_kode, 'partinavn':self.parti_navn, 'valgår': self.valgår}
        data = json.dumps(fortegnelse)
        return json.loads(fortegnelse)

venstre=Parti('V','Venstre')
venstre.skriv()
venstre.leggTilÅr(2013)
venstre.leggTilÅr(2017)
venstre.skriv()
venstre.json()


Venstre (V) valgår:[]
Venstre (V) valgår:[2013, 2017]


'{"partikode": "V", "partinavn": "Venstre", "valg\\u00e5r": [2013, 2017]}'

In [27]:
def emnekode(søk_tittel):
    for (emne_kode, studiepoeng, emnetittel, forkunnskaper) in emner:
        if søk_tittel == emnetittel:
            return emne_kode
    return 'ukjent emne'

emner = [('INFO132', 10, 'Innføring i programmering', []), ('DATA3', 5, 'Algoritmer og datastrukturer', ['INFO132'])]
emnekode('Algoritmer og datastrukturer')


'DATA3'

In [ ]:
eksamener = {
    'Liv': {('DATA1', 'B'), ('KI1', 'C'), ('MAT1', 'C'), ('DATA3', 'C'), ('MAT5', 'B')},
    'Per': {('MAT1', 'C'), ('ØKO1', 'A'), ('KI1', 'A'), ('MAT4', 'C'), ('MAT2', 'B')},
    'Gro': {('ØKO1', 'B'), ('MAT1', 'A'), ('ØKO2', 'A'), ('MAT2', 'A'), ('MAT4', 'A')},
    'Jan': {('DATA1', 'B'), ('MAT1', 'B'), ('MAT5', 'D'), ('DATA3', 'A'), ('KI1', 'B')},
    'Ida': {('MAT5', 'C'), ('MAT1', 'A'), ('DATA1', 'D'), ('KI1', 'B'), ('MAT10', 'B')},
    'Dag': {('DATA2', 'A'), ('MAT1', 'B'), ('DATA3', 'C'), ('MAT5', 'A'), ('KI1', 'D')},
    'Ine': {('ØKO1', 'C'), ('KI1', 'C'), ('DATA1', 'A'), ('MAT1', 'D')},
    'Jonas': {('KI1', 'C')}
}
emner = [('INFO132', 10, 'Innføring i programmering', []), ('DATA3', 5, 'Algoritmer og datastrukturer', ['INFO132'])]




In [30]:
eksamener = {
    'Liv': {('DATA1', 'B'), ('KI1', 'C'), ('MAT1', 'C'), ('DATA3', 'C'), ('MAT5', 'B')},
    'Per': {('MAT1', 'C'), ('ØKO1', 'A'), ('KI1', 'A'), ('MAT4', 'C'), ('MAT2', 'B')},
    'Gro': {('ØKO1', 'B'), ('MAT1', 'A'), ('ØKO2', 'A'), ('MAT2', 'A'), ('MAT4', 'A')},
    'Jan': {('DATA1', 'B'), ('MAT1', 'B'), ('MAT5', 'D'), ('DATA3', 'A'), ('KI1', 'B')},
    'Ida': {('MAT5', 'C'), ('MAT1', 'A'), ('DATA1', 'D'), ('KI1', 'B'), ('MAT10', 'B')},
    'Dag': {('DATA2', 'A'), ('MAT1', 'B'), ('DATA3', 'C'), ('MAT5', 'A'), ('KI1', 'D')},
    'Ine': {('ØKO1', 'C'), ('KI1', 'C'), ('DATA1', 'A'), ('MAT1', 'D')},
    'Jonas': {('KI1', 'C')}
}
emner = [('INFO132', 10, 'Innføring i programmering', []), ('DATA3', 5, 'Algoritmer og datastrukturer', ['INFO132'])]


def nyEksamen(person, emne, karakter):
    #legger til evt. ny person i eksamener-fortegnelsen
    if person not in eksamener:
        eksamener[person] = set((emne, karakter))
        return None #retunerer None siden det er unødvendige at resten av funksjonen utføres.
   
   #fjerner evt. gammel karakter, discard for å unngå havari
    eksamener[person].discard((emne, karakter))
    eksamener[person].add((emne, karakter)) #legger til ny karakter

nyEksamen('Per','MAT2','B') 
nyEksamen('Gro', 'MAT1', 'A') 
nyEksamen('Jonas', 'KI1', 'C')
        
print(eksamener)


{'Liv': {('MAT5', 'B'), ('DATA1', 'B'), ('MAT1', 'C'), ('KI1', 'C'), ('DATA3', 'C')}, 'Per': {('ØKO1', 'A'), ('MAT2', 'B'), ('MAT1', 'C'), ('KI1', 'A'), ('MAT4', 'C')}, 'Gro': {('ØKO1', 'B'), ('ØKO2', 'A'), ('MAT2', 'A'), ('MAT4', 'A'), ('MAT1', 'A')}, 'Jan': {('KI1', 'B'), ('DATA3', 'A'), ('MAT1', 'B'), ('DATA1', 'B'), ('MAT5', 'D')}, 'Ida': {('MAT10', 'B'), ('KI1', 'B'), ('MAT1', 'A'), ('DATA1', 'D'), ('MAT5', 'C')}, 'Dag': {('MAT1', 'B'), ('KI1', 'D'), ('DATA2', 'A'), ('MAT5', 'A'), ('DATA3', 'C')}, 'Ine': {('ØKO1', 'C'), ('MAT1', 'D'), ('KI1', 'C'), ('DATA1', 'A')}, 'Jonas': {('KI1', 'C')}}


In [8]:
#bruker emnekode-funksjonen fra oppg.3a, men fjerner siste programsetning
#for at funksjonen skal retunere 'None' om emnekode ikke finnes.
def emnekode(søk_tittel):
    for (emne_kode, studiepoeng, emnetittel, forkunnskaper) in emner:
        if søk_tittel == emnetittel:
            return emne_kode
        
def karakterer(person, emnetittel):
    #retunerer melding dersom emnekode ikke finnes.
    emne_kode = emnekode(emnetittel)
    for (emne, karakter) in eksamener[person]:
        if emne == emne_kode:
            return karakter
    return f'{person} har ikke tatt {emnetittel}'

eksamener = {
    'Liv': {('DATA1', 'B'), ('KI1', 'C'), ('MAT1', 'C'), ('DATA3', 'C'), ('MAT5', 'B')},
    'Per': {('MAT1', 'C'), ('ØKO1', 'A'), ('KI1', 'A'), ('MAT4', 'C'), ('MAT2', 'B')},
    'Gro': {('ØKO1', 'B'), ('MAT1', 'A'), ('ØKO2', 'A'), ('MAT2', 'A'), ('MAT4', 'A')},
    'Jan': {('DATA1', 'B'), ('MAT1', 'B'), ('MAT5', 'D'), ('DATA3', 'A'), ('KI1', 'B')},
    'Ida': {('MAT5', 'C'), ('MAT1', 'A'), ('DATA1', 'D'), ('KI1', 'B'), ('MAT10', 'B')},
    'Dag': {('DATA2', 'A'), ('MAT1', 'B'), ('DATA3', 'C'), ('MAT5', 'A'), ('KI1', 'D')},
    'Ine': {('ØKO1', 'C'), ('KI1', 'C'), ('DATA1', 'A'), ('MAT1', 'D')},
    'Jonas': {('KI1', 'C')}
}
emner = [('DATA1', 10, 'Innføring i programmering', []), ('DATA3', 5, 'Algoritmer og datastrukturer', ['INFO132'])]

karakterer('Liv','Innføring i programmering')

'B'

1a) forklaring konkatenering -> lister med multiplikasjon er ikke konkatinering
1c) 3 feil: [2,3,4] in [1,2,3,4,5], er json-filer er tekst-filer, 'requests er en modul for serialisering av objekter'
2d) blant annet feil med dumps og loads
3c) feil med return i betinget utførelse

Oppgave 1
a)  
    1.  ?/1.5  upresist i setninger: ...bruke konkatenering til å plusse (+) eller gange(*) blant annet lister.
    2.  1.5/1.5     
    3.  1.5/1.5  
    4.  1.5/1.5   

b)
    4/4

c) 
    1/1
    1/1
    0/1 feil i spørsmål:[2,3,4] in [1,2,3,4,5]
    0/1 feil i spørsmål er json-filer er tekst-filer
    1/1
    0/1 feil i spørsmål: 'requests er en modul for serialisering av objekter'
    1/1
    1/1
    1/1

d)
    1/1
    1/1
    1/1
    1/1
    1/1
    1/1
    1/1
    1/1
    1/1
    1/1

e)
    ?/10 mulig feil: (mengde -> set), (fil->open), (.get->stream), (.remove->.pop), (JSON->.dump), (while->continue), (klasse->.__init__), (GUI->.mainloop), (API->endepunkt), (fortegnelse->.items)


Oppgave 2
a)
    6/6 

b)
    10/10

c)
    10/10

d)
    ?/10 feil med dump og load i json metoden.

Oppgave 3
a)
    6/6

b)
    10/10

c)
    ?/10 feil i programsetning: emne_kode = emnekode(emnetittel) if emnekode(emnetittel) else return 'Emnetittel ikke funnet'

d)
    12/12


In [ ]:
er det ein feil i koden her? 
Beskrivelse F:  f(x) returnerer den fullstendige listen av forkunnskaper man må ha for å ta emnet x.  
(Merk at forkunnskapene som er registrert i emner-listen kan selv ha forkunnskaper som også skal med i resultatet her)

Eksempel
>>>f('ØKO2')
['MAT4', 'MAT2', 'MAT1', 'ØKO1'] 
def f2(a):
    for (b,c,d,e) in emner:
        if b==a:
            h=set(e)
            break
    i=set(e)
    while i != set():
         j=i.pop()
         for (k,l,m,n) in emner:
            if k==j:
                i.update(set(n))
                h.add(j)
    return list(h)